---
title: "Interconnected lineage trajectories link conventional and NK-like exhausted CD8+ T cells beneficial in T1D"
output: pdf_document
---

 - data reproduction of Erin Witkop's ATAC-seq
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(ggplot2)
library(dplyr)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Create reference genome fasta

```{r}

library(BSgenome.Hsapiens.UCSC.hg38)
mainChromosomes <- paste0("chr",c(1:21,"X","Y","M"))
mainChrSeq <- lapply(mainChromosomes,
                     function(x)BSgenome.Hsapiens.UCSC.hg38[[x]])
names(mainChrSeq) <- mainChromosomes
mainChrSeqSet <- DNAStringSet(mainChrSeq)
writeXStringSet(mainChrSeqSet,
                "BSgenome.Hsapiens.UCSC.hg38.mainChrs.fa")

```

# Create Rsubread index to allow alignment (or load if already created)

```{r}

library(Rsubread)
buildindex("../saved_data/BSgenome.Hsapiens.UCSC.hg38.mainChrs",
           "BSgenome.Hsapiens.UCSC.hg38.mainChrs.fa",
           indexSplit = TRUE,
           memory = 1000,
           saveDB = TRUE)

# loadDb("../saved_data/BSgenome.Hsapiens.UCSC.hg38.mainChrs")

```

# Read in fastq files

```{r}

read1 <- "../raw_data/"
read2 <- "ATAC_Data/ATAC_FQs/SRR891269_2.fastq.gz"

require(ShortRead)
read1 <- readFastq("data/ATACSample_r1.fastq.gz")
read2 <- readFastq("data/ATACSample_r2.fastq.gz")
id(read1)[1:2]

```

# Align reads

```{r}

align("BSgenome.Hsapiens.UCSC.hg38.mainChrs",
      readfile1=read1,readfile2=read2,
      output_file = "ATAC_50K_2.bam",
      nthreads=2,type=1,
      unique=TRUE,maxFragLength = 2000)

library(Rbowtie2)
bowtie2_build(references="BSgenome.Hsapiens.UCSC.hg38.mainChrs.fa", 
              bt2Index="BSgenome.Hsapiens.UCSC.hg38.mainChrs_bowtie2")

```

# Decompress fastq.gz

```{r}

gunzip("ATAC_Data/ATAC_FQs/SRR891269_1.fastq.gz")
gunzip("ATAC_Data/ATAC_FQs/SRR891269_2.fastq.gz")

```

# Create Rbowtie2 index to allow alignment of fastq to genome with bowtie

```{r}

library(Rsamtools)
bowtie2(bt2Index = "BSgenome.Hsapiens.UCSC.hg38.mainChrs_bowtie2",
          samOutput = "ATAC_50K_2_bowtie2.sam",
          seq1 = "ATAC_Data/ATAC_FQs/SRR891269_1.fastq",
          seq1 = "ATAC_Data/ATAC_FQs/SRR891269_2.fastq"
        )
asBam("ATAC_50K_2_bowtie2.sam")

```

# Recompress the FASTQs and remove the SAM files with the unlink() function

```{r}


```

# Sort and index BAM file for use with external tools

```{r}

library(Rsamtools)

# sort data by sequence order
sortedBAM <- file.path(dirname(outBAM),
                       paste0("Sorted_",basename(outBAM))
                       )
sortBam(outBAM,gsub("\\.bam","",basename(sortedBAM)))

# index file to allow rapid access of genomic locations by other tools
indexBam(sortedBAM)

```

# Check distribution of reads

```{r}

library(Rsamtools)
mappedReads <- idxstatsBam(sortedBAM)

```

# Plot distribution of reads across chr

```{r}

library(ggplot2)
ggplot(mappedReads, aes(seqnames, mapped, fill = seqnames)) + geom_bar(stat = "identity") +
    coord_flip()

```