---
title: "Interconnected lineage trajectories link conventional and NK-like exhausted CD8+ T cells beneficial in T1D"
output: pdf_document
---

 - data reproduction of Erin Witkop's ATAC-seq
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(Rsamtools)
library(ggplot2)
library(dplyr)
library(GenomicAlignments)
library(Rsubread)
library(ATACseqQC)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Create list of sorted .bam files

```{r}

bam_dir <- "/mnt/bioinformatics/pipeline/Illumina/220802_VH00126_203_AAAMVKNHV/Project_P452-3Processed_globus_220807/shiftedAlignments/"

bam_files <- list.files(bam_dir, pattern = "\\_sorted.bam$", full.names = TRUE)

```

# Check mapping percent for all sorted .bam files

```{r}

all_map_percs <- data.frame()

for (bam_file in bam_files) {
  map_perc <- propmapped(bam_file)
  map_perc$file <- basename(bam_file)
  map_perc$libid <- sub("lib([0-9]+)_.*", "\\1", map_perc$file)
  all_map_percs <- rbind(all_map_percs, map_perc)
}

all_map_percs

```

# Look on which (expected) chromosomes all sorted .bam files aligned

```{r}

all_mapped_reads <- data.frame()

for (bam_file in bam_files) {
  # some reads mapped to alternative chr, but less commonly
  mapped_reads <- idxstatsBam(bam_file) %>% filter(seqnames %in% c(1:22, "X", "Y", "MT"))
  mapped_reads$file <- basename(bam_file)
  mapped_reads$libid <- sub("lib([0-9]+)_.*", "\\1", mapped_reads$file)
  all_mapped_reads <- rbind(all_mapped_reads, mapped_reads)
}

# lots of files, make 2 plots to see both here, all pretty similar (most reads map to mitochondrial genome)

ggplot(all_mapped_reads %>% filter(libid < 82449), aes(seqnames, mapped, fill = seqnames)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  facet_wrap(~ libid, ncol = 6) +
  theme(legend.position = "none",
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6)) +
  labs(x = "Chromosome", y = "Mapped counts")

p <- ggplot(all_mapped_reads %>% filter(libid >= 82449), aes(seqnames, mapped, fill = seqnames)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  facet_wrap(~ libid, ncol = 6) +
  theme(legend.position = "none",
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6)) +
  labs(x = "Chromosome", y = "Mapped counts")

p

ggsave("../figures/chr_map.svg", p)

```

# Plot percent mitochondrial reads for each sorted .bam file

```{r}

perc_mt <- all_mapped_reads %>%
  group_by(libid) %>%
  summarize(mt_perc = 100*mapped[seqnames == "MT"] / sum(mapped))

perc_mt

p <- ggplot(perc_mt, aes(libid, mt_perc, fill = libid)) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none",
        axis.text.x = element_text(size = 8, angle = 45, hjust = 1)) +
  labs(x = "Lib ID", y = "Mapped counts to MT (%)")

p

ggsave("../figures/mt_map_perc.svg", p)

```

# Extract (paired) read info from chr1 from 3 sorted .bam files

 - crashes if trying to do for all bam files

```{r}

# hg38 chr sizes (from UCSC genome browser)
# chr20:1-64,444,167
# chrX:1-156,040,895
# chr1:1-248,956,422

read_chr_1 <- function(bam_file_path) {
  chr1_reads <- readGAlignmentPairs(bam_file_path, 
                                    param = ScanBamParam(mapqFilter = 1, 
                                                          flag = scanBamFlag(isPaired = TRUE, isProperPair = TRUE), 
                                                          what = c("qname", "mapq", "isize"), 
                                                          which = GRanges("1", IRanges(1, 248956422))))
  mcols(chr1_reads)$libid <- sub("lib([0-9]+)_.*", "\\1", basename(bam_file_path))
  
  return(chr1_reads)
}

chr1_list <- lapply(bam_files[1:3], read_chr_1)

```

```{r}
chr1_list
```

# Plot read fragment length distribution for chr 1 mapping reads from 3 sorted .bam files

```{r}

# need to update this to work on chr1_list each object of which is now like below chr1_reads

chr1_reads_1 <- first(chr1_reads)
insertSizes <- abs(elementMetadata(chr1_reads_1)$isize)

fragLenSizes <- table(insertSizes)

toPlot <- data.frame(InsertSize = as.numeric(names(fragLenSizes)), Count = as.numeric(fragLenSizes))

# filter for insert sizes < 2 kb because there are some that are wildly longer messing up plot...
p <- ggplot(toPlot %>% filter(InsertSize < 2000), aes(x = InsertSize, y = Count)) +
  geom_line() +
  scale_y_continuous(trans = "log2") +
  geom_vline(xintercept = c(180, 247), colour = "red") +
  geom_vline(xintercept = c(315, 437), colour = "darkblue") +
  geom_vline(xintercept = c(100), colour = "darkgreen")

p

# annotate our nucleosome free (< 100bp), mononucleosome (180bp-247bp) and dinucleosome (315-437) lengths

# generally looks ok except for rare, long inserts (mapping artifacts and/or underdigestion due to too many cells)

```

# Look at MapQ scores for 1 sorted .bam file for chr 1

```{r}

read1MapQ <- mcols(chr1_reads_1)$mapq
read2MapQ <- mcols(second(chr1_reads))$mapq

read1MapQFreqs <- table(read1MapQ)
read2MapQFreqs <- table(read2MapQ)

library(ggplot2)
toPlot <- data.frame(MapQ = c(names(read1MapQFreqs), names(read2MapQFreqs)), Frequency = c(read1MapQFreqs,
    read2MapQFreqs), Read = c(rep("Read1", length(read1MapQFreqs)), rep("Read2",
    length(read2MapQFreqs))))
toPlot$MapQ <- factor(toPlot$MapQ, levels = unique(sort(as.numeric(toPlot$MapQ))))
ggplot(toPlot, aes(x = MapQ, y = Frequency, fill = MapQ)) + geom_bar(stat = "identity") +
    facet_grid(~Read)

# The quality of the alignment, or how well a read matches a specific region of the genome, is reported as a MAPQ score. The highest MAPQ value (255) goes to reads that are unique for a single region, while the lowest (0) goes to those that have a lot of matching regions

# so most reads here are uniquely mapping!

```
 
# More QC

```{r}

ATACQC <- bamQC("/mnt/bioinformatics/pipeline/Illumina/220802_VH00126_203_AAAMVKNHV/Project_P452-3Processed_globus_220807/shiftedAlignments/lib82438_AAAMVKNHV_star_alignments_shifted_sorted.bam")
names(ATACQC)

# PCR bottleneck coefficients identify PCR bias/overamplification in library prep
# PCR bottleneck coefficient 1 = positions with only 1 read compared to other positions with >= 1 read, < 0.7 bad bottlenecking, 0.7-0.9 ok, > 0.9 good no bottleneck

ATACQC$PCRbottleneckCoefficient_1

# coefficient 2 is positions with 1 read compared to those with 2 reads, > 3 good, 1-3 ok, < 1 bad

ATACQC$PCRbottleneckCoefficient_2

```

# Investigate total read numbers

# Investigate percent GC

# Investigate percent correct strand reads

# Investigate percent duplication

# Investigate percent over-represented sequences

# Investigate median CV coverage

# Investigate 3' and 5' bias

# Plot encichment around TF start site
 - run P452_3_TSSe_biometal.R to calculate transcription factor start site enrichment

BiocManager::install("soGGi")
library(soGGi)

library(TxDb.Hsapiens.UCSC.hg38.knownGene)
TxDb.Hsapiens.UCSC.hg38.knownGene

# Notes
 - Erin code: https://github.com/BenaroyaResearch/Witkop_T1DAL_2023/tree/main
 - Stephan advice: use /shiftedAlignments as bamfiles; /peaks contains peak files