---
title: "Interconnected lineage trajectories link conventional and NK-like exhausted CD8+ T cells beneficial in T1D"
output: pdf_document
---

 - data reproduction of Erin Witkop's ATAC-seq
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(Rsamtools)
library(ggplot2)
library(dplyr)
library(GenomicAlignments)
library(Rsubread)
library(ATACseqQC)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(soGGi)
library(BiocParallel)
register(SerialParam())
library(DiffBind)
library(ChIPseeker)
library(org.Hs.eg.db)
library(tidyr)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Read in RDS

```{r}

model_norm <- readRDS("../saved_data/normalized_data.rds")

```

# Annotate peaks

```{r}

# get peaks out of dba object
peaks <- dba.peakset(model_norm, bRetrieve = TRUE)
# could use dba.report() for just differentially bound peaks

# add chr to seqnames to get annotatePeak function working
new_seqnames <- paste0("chr", seqlevels(peaks))
new_names_vector <- setNames(new_seqnames, seqlevels(peaks))
peaks_with_new_seq_level <- renameSeqlevels(peaks, new_names_vector)

txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

bed.annot = annotatePeak(peaks_with_new_seq_level, tssRegion=c(-3000, 3000),TxDb=txdb, annoDb="org.Hs.eg.db")

annot_peaks=as.data.frame(bed.annot)

annot_peaks

```

# Combine gene annotations with diffbind results

```{r}

anno_data <- annot_peaks %>%
  dplyr::select(seqnames, start, end, strand, width, geneChr, geneStart, geneEnd, geneLength, geneStrand, geneId, transcriptId, distanceToTSS, ENSEMBL, SYMBOL, GENENAME) %>%
  mutate(seqnames = gsub("chr", "", seqnames))

contrast_1_report <- dba.report(model_norm, method=DBA_EDGER,
                                              contrast = 1,  
                                              bUsePval = FALSE,
                                              th=1) %>%
                     as_tibble() %>%
                     left_join(anno_data, by = c("seqnames", "start", "end", "width", "strand"))

contrast_2_report <- dba.report(model_norm, method=DBA_EDGER,
                                              contrast = 2,  
                                              bUsePval = FALSE,
                                              th=1) %>%
                     as_tibble() %>%
                     left_join(anno_data, by = c("seqnames", "start", "end", "width", "strand"))

contrast_3_report <- dba.report(model_norm, method=DBA_EDGER,
                                              contrast = 3,  
                                              bUsePval = FALSE,
                                              th=1) %>%
                     as_tibble() %>%
                     left_join(anno_data, by = c("seqnames", "start", "end", "width", "strand"))

```

# Mean coverage plots for specific genes for each of 3 cell type groups

 - NCR1, KLRF1, KIR3DL2, KIR2DL3

```{r}



```