---
title: "Identification of Pathogenic Immune Cell Subsets Associated with Checkpoint Inhibitor-induced Myocarditis"
output: pdf_document
---

 - data reproduction of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9397491/
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(Seurat)
library(Matrix)
library(ggplot2)
library(dplyr)
library(tidyr)
library(stringr)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Read in non-control TCR-seq data

```{r}

data_directory <- "../raw_data"

# Identify the pattern to extract patient IDs using regular expressions
pattern <- "GSM\\d+_([A-Z\\d\\-]+)-filtered_contig_annotations"

# Get a list of CSV files in the directory
csv_files <- list.files(data_directory, pattern = ".csv", full.names = TRUE)

# Initialize an empty list to store data frames
data_frames <- list()

# Loop through the list of CSV files
for (file in csv_files) {
  # Extract patient ID from the file name using regex
  patient_id <- gsub("GSM(\\d+)_(.*?)\\-filtered_contig_annotations.csv", "\\2", basename(file))
  
  # Read the CSV file into a data frame
  df <- read.csv(file)
  
  # Add a column for patient ID
  df$Patient_ID <- patient_id
  
  # Append the data frame to the list
  data_frames[[patient_id]] <- df
  
}

# Combine all data frames into a single data frame
combined_tcr_seq <- bind_rows(data_frames)

```

# Read in dehashed RDS

```{r}

dehashed <- readRDS("../saved_data/dehashed.rds")

```

# Dehash TCRseq data, lose some barcodes in process (0-lets, multiplets, no hash marker)

```{r}

A6 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$A6)) %>%
  mutate(Patient_ID = "A6")

A9 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$A9)) %>%
  mutate(Patient_ID = "A9")

B13 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$B13)) %>%
  mutate(Patient_ID = "B13")

B9 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$B9)) %>%
  mutate(Patient_ID = "B9")

A8 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A8-B11"),
         barcode %in% colnames(dehashed$A8)) %>%
  mutate(Patient_ID = "A8")

B11 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A8-B11"),
         barcode %in% colnames(dehashed$B11)) %>%
  mutate(Patient_ID = "B11")

B12 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("B12-MCE6-MCE7"),
         barcode %in% colnames(dehashed$B12)) %>%
  mutate(Patient_ID = "B12")

MCE6 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("B12-MCE6-MCE7"),
         barcode %in% colnames(dehashed$MCE6)) %>%
  mutate(Patient_ID = "MCE6")

MCE7 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("B12-MCE6-MCE7"),
         barcode %in% colnames(dehashed$MCE7)) %>%
  mutate(Patient_ID = "MCE7")

A10 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$A10)) %>%
  mutate(Patient_ID = "A10")

B10 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$B10)) %>%
  mutate(Patient_ID = "B10")

MCE12 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$MCE12)) %>%
  mutate(Patient_ID = "MCE12")

MCE13 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$MCE13)) %>%
  mutate(Patient_ID = "MCE13")

A5 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A5-B14"),
         barcode %in% colnames(dehashed$A5)) %>%
  mutate(Patient_ID = "A5")

B14 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A5-B14"),
         barcode %in% colnames(dehashed$B14)) %>%
  mutate(Patient_ID = "B14")

dehashed_combined_tcr_seq <- combined_tcr_seq %>%
  filter(!(Patient_ID %in% c("A5-B14", "A10-B10-MCE12-MCE13", "B12-MCE6-MCE7", "A6-A9-B13-B9", "A8-B11")),
         productive != "False")

# Define a list of data frames for each patient or barcode
patient_dataframes <- list(A6, A9, B13, B9, A8, B11, B12, MCE6, MCE7, A10, B10, MCE12, MCE13, A5, B14)

# Concatenate the data frames into a single data frame
dehashed_combined_tcr_seq <- bind_rows(dehashed_combined_tcr_seq, patient_dataframes)

dehashed_combined_tcr_seq <- dehashed_combined_tcr_seq %>%
  mutate(irAE = case_when(
    startsWith(Patient_ID, "A") ~ "No",
    startsWith(Patient_ID, "B") ~ "Yes",
    startsWith(Patient_ID, "M") ~ "Yes"))

dehashed_combined_tcr_seq

```

# Prep for stitchr

```{r}
stitchr_data <- dehashed_combined_tcr_seq %>%
  filter(chain %in% c("TRA", "TRB")) %>%
  group_by(barcode) %>%
  filter(sum(chain == "TRA") == 1 & sum(chain == "TRB") == 1) %>%
  ungroup() %>%
  select(barcode, v_gene, d_gene, j_gene, c_gene, chain, cdr3_nt) %>%
  rename(TCR_name = barcode,
         CDR3 = `cdr3_nt`,
         V = `v_gene`,
         J = `j_gene`,
         C = `c_gene`) %>%
  pivot_wider(names_from = chain, values_from = c(V, J, C, CDR3)) %>%
  # rename, reorder, create cols for stitchr
  rename(TRAV = V_TRA, TRAJ = J_TRA, TRBV = V_TRB, TRBJ = J_TRB, TRAC = C_TRA, TRBC = C_TRB, TRA_CDR3 = CDR3_TRA, TRB_CDR3 = CDR3_TRB) %>%
  select(TCR_name, TRAV, TRAJ, TRA_CDR3, TRBV, TRBJ, TRB_CDR3, TRAC, TRBC) %>%
  filter_all(all_vars(!is.na(.))) %>%
  filter_all(all_vars(. != "None")) %>%
  mutate(TRA_leader	= NA,
         TRB_leader = NA,
         Linker = NA,
         Link_order = NA,
         TRA_5_prime_seq = NA,
         TRA_3_prime_seq = NA,
         TRB_5_prime_seq = NA,
         TRB_3_prime_seq = NA)

stitchr_data

# 32,089 rows

write.table(stitchr_data, file = "../saved_data/tcrseq_for_stitchr.tsv", sep = "\t", quote = FALSE, row.names = FALSE)

```

# Read in stitchr output

```{r}

stitchr_output <- read.delim("../saved_data/stitchr_out.tsv", header = TRUE, sep = "\t")

# Columns that need further separation by space
columns_to_split_by_space <- c("TRA_nt", "TRA_aa", "TRB_nt", "TRB_aa", "Linked_nt")

# Split the specified columns by space
for (col_name in columns_to_split_by_space) {
  stitchr_output[[col_name]] <- sapply(stitchr_output[[col_name]], function(x) unlist(strsplit(x, " ")))
}

# Assign the column names
colnames(stitchr_output) <- c("TCR_name", "TRA_nt", "TRB_nt", "TRA_aa", "TRB_aa", "TRAV", "TRAJ", "TRA_CDR3", "TRBV", "TRBJ",
                    "TRB_CDR3", "TRAC", "TRBC", "TRA_leader", "TRB_leader", "Linker", "Link_order", 
                    "TRA_5_prime_seq", "TRA_3_prime_seq", "TRB_5_prime_seq", "TRB_3_prime_seq", "Linked_nt", 
                    "Linked_aa", "Warnings.Errors")

# Define a function to replace with NA if length is 0
replace_empty_with_na <- function(x) {
  if (length(x) == 0) {
    return(NA)
  } else {
    return(x)
  }
}

# Replace values in the specified columns with real NA
cols_to_update <- c("TRA_nt", "TRB_nt", "TRA_aa", "TRB_aa", "Linked_nt")

for (col_name in cols_to_update) {
  stitchr_output[[col_name]] <- lapply(stitchr_output[[col_name]], replace_empty_with_na)
}

stitchr_output$TRA_nt <- sapply(stitchr_output$TRA_nt, as.character)
stitchr_output$TRA_aa <- sapply(stitchr_output$TRA_aa, as.character)
stitchr_output$TRB_nt <- sapply(stitchr_output$TRB_nt, as.character)
stitchr_output$TRB_aa <- sapply(stitchr_output$TRB_aa, as.character)
stitchr_output$Linked_nt <- sapply(stitchr_output$Linked_nt, as.character)

```

# Analyze stitchr errors. "Unable to locate N terminus of CDR3 in V gene correctly. Please ensure sequence plausibility" is main error for TRA/TRB

```{r}

filtered_results_list <- list()

for (i in 1:nrow(stitchr_output)) {
  # Split the row using the specified pattern and extract the relevant elements
  split_result <- strsplit(stitchr_output$Warnings.Errors[i], "\\(TRA\\)|\\(TRB\\)")[[1]]
  filtered_list <- split_result[grep("Cannot stitch", split_result)]
  
  # Store the filtered list for this row in the results list
  filtered_results_list[[i]] <- filtered_list
}

your_df <- data.frame(Column1 = unlist(unlist(filtered_results_list)))

# Optionally, rename the column
colnames(your_df) <- c("errors")

count_df <- your_df %>%
  group_by(errors) %>%
  summarize(count = n()) %>%
  arrange(-count)

count_df

# try to see if there are commonalities to V genes/CDR3s for issue-causing rows

errors <- stitchr_output %>%
  mutate(common_error = if_else(str_detect(Warnings.Errors, "Unable to locate N terminus of CDR3 in V gene correctly"), "Yes", "No")) %>%
  arrange(common_error) %>%
  select(TRA_CDR3, TRAV, TRB_CDR3, TRBV, common_error)

```

# Prep IGoR input

```{r}

stitchr_output_alpha_df <- stitchr_output %>%
  filter(!(is.na(TRA_nt)))

stitchr_output_alpha_df
# 9,475 rows

write.csv(stitchr_output_alpha_df, file = "../saved_data/igor_alpha_input.csv", row.names = FALSE)

stitchr_output_beta_df <- stitchr_output %>%
  filter(!(is.na(TRB_nt)))

stitchr_output_beta_df
# 22,053 rows

write.csv(stitchr_output_beta_df, file = "../saved_data/igor_beta_input.csv", row.names = FALSE)

```

# Read in IGoR output

```{r}

igor_alpha <- read.csv("../saved_data/igor_output_alpha.csv") %>%
  rename(barcode = TCR_name) %>%
  left_join(dehashed_combined_tcr_seq, by = "barcode") %>%
  mutate(patient_group = case_when(
    startsWith(Patient_ID, "A") ~ "ici_no_irae",
    startsWith(Patient_ID, "B") ~ "ici_non_myocarditis_irae",
    startsWith(Patient_ID, "M") ~ "ici_myocarditis_irae")) %>%
  mutate(irAE = case_when(
    startsWith(Patient_ID, "A") ~ "No",
    startsWith(Patient_ID, "B") ~ "Yes",
    startsWith(Patient_ID, "M") ~ "Yes")) %>%
  select(patient_group, irAE, chain, pgen, barcode, cdr3) %>%
  filter(chain == "TRA",
         pgen != "NaN")

igor_combined <- read.csv("../saved_data/igor_output_beta.csv") %>%
  rename(barcode = TCR_name) %>%
  left_join(dehashed_combined_tcr_seq, by = "barcode") %>%
  mutate(patient_group = case_when(
    startsWith(Patient_ID, "A") ~ "ici_no_irae",
    startsWith(Patient_ID, "B") ~ "ici_non_myocarditis_irae",
    startsWith(Patient_ID, "M") ~ "ici_myocarditis_irae")) %>%
  mutate(irAE = case_when(
    startsWith(Patient_ID, "A") ~ "No",
    startsWith(Patient_ID, "B") ~ "Yes",
    startsWith(Patient_ID, "M") ~ "Yes")) %>%
  select(Patient_ID, patient_group, irAE, chain, pgen, barcode, cdr3) %>%
  filter(chain == "TRB",
         pgen != "NaN") %>%
  full_join(igor_alpha, by = c("barcode", "chain", "patient_group", "pgen", "cdr3", "irAE"))

igor_combined

```

# Plot pgen scores across all TCRs by irAE development

```{r}

ggplot(igor_combined, aes(x = chain, y = -log10(pgen), fill = irAE)) +
  geom_boxplot() +
  labs(x = "Chain", y = "-log10(pgen)")

# Filter the data for statistical test
test_data <- igor_combined %>%
  filter(chain == "TRA")

# Calculate the Wilcoxon rank-sum test statistic and p-value
wilcox_result <- wilcox.test(test_data$pgen ~ test_data$irAE)

# Print the results
print(wilcox_result)

```

# Prep IMGT/HighV-Quest input

```{r}

fasta_prep <- stitchr_output %>%
  rename(barcode = TCR_name) %>%
  left_join(dehashed_combined_tcr_seq, by = "barcode")

fasta_df_alpha <- fasta_prep %>%
  filter(!(is.na(TRA_nt)),
         chain == "TRA") %>%
  mutate(barcode = paste(barcode, Patient_ID, sep = "_")) %>%
  ungroup()

# Function to convert a data frame to FASTA format
df_to_fasta_alpha <- function(df, output_file) {
  # Open/create the output file for writing
  fasta_file <- file(output_file, "w")
  
  # Iterate through rows of the data frame
  for (i in 1:nrow(df)) {
    # Combine name columns and add a suffix numbering
    label <- df$barcode[i]

    # Write the sequence and label to the FASTA file
    writeLines(paste(">", label), con = fasta_file)
    writeLines(df$TRA_nt[i], con = fasta_file)
  }
  
  # Close the output file
  close(fasta_file)
}

# Call the function to convert the data frame to FASTA format
df_to_fasta_alpha(fasta_df_alpha, "../saved_data/vdjseq_alpha.fasta")

fasta_df_beta <- fasta_prep %>%
  filter(!(is.na(TRB_nt)),
         chain == "TRB") %>%
  mutate(barcode = paste(barcode, Patient_ID, sep = "_")) %>%
  ungroup()

# Function to convert a data frame to FASTA format
df_to_fasta_beta <- function(df, output_file) {
  # Open/create the output file for writing
  fasta_file <- file(output_file, "w")
  
  # Iterate through rows of the data frame
  for (i in 1:nrow(df)) {
    # Combine name columns and add a suffix numbering
    label <- df$barcode[i]

    # Write the sequence and label to the FASTA file
    writeLines(paste(">", label), con = fasta_file)
    writeLines(df$TRB_nt[i], con = fasta_file)
  }
  
  # Close the output file
  close(fasta_file)
}

# Call the function to convert the data frame to FASTA format
df_to_fasta_beta(fasta_df_beta, "../saved_data/vdjseq_beta.fasta")

```

# Analyze IMGT/HighV-Quest output 8: V region nt mutation statistics

```{r}

imgt_df_8_alpha <- read.table("../saved_data/myocarditis_dehashed_alpha/8_V-REGION-nt-mutation-statistics.txt", header = TRUE, sep = "\t") %>%
  select(-Sequence.number) %>%
  rename(Barcode = Sequence.ID)

colnames(imgt_df_8_alpha)[colnames(imgt_df_8_alpha) != "Barcode"] <- paste0(colnames(imgt_df_8_alpha)[colnames(imgt_df_8_alpha) != "Barcode"], "_alpha")

imgt_df_8_beta <- read.table("../saved_data/myocarditis_dehashed_beta/8_V-REGION-nt-mutation-statistics.txt", header = TRUE, sep = "\t") %>%
  select(-Sequence.number) %>%
  rename(Barcode = Sequence.ID)

colnames(imgt_df_8_beta)[colnames(imgt_df_8_beta) != "Barcode"] <- paste0(colnames(imgt_df_8_beta)[colnames(imgt_df_8_beta) != "Barcode"], "_beta")
  
imgt_df_8 <- imgt_df_8_alpha %>%
  full_join(imgt_df_8_beta, by = "Barcode") %>%
  # too many columns, focus on CDR3 ones for now
  select(matches("^CDR3|^Barcode"))

# join with patient ID to get patient group --> plots

imgt_df_8

```

# Prep hydrophobicity calculator

```{r}

# Define the amino acids and their corresponding Eisenberg hydrophobicity values
amino_acids <- c("A", "R", "N", "D", "C", "Q", "E", "G", "H", "I", "L", "K", "M", "F", "P", "S", "T", "W", "Y", "V")
hydrophobicity_values <- c(0.620, -2.530, -0.780, -0.900, 0.290, -0.850, -0.740, 0.480, -0.400, 1.380, 1.060, -1.500, 0.640, 1.190, 0.120, -0.180, -0.050, 0.81, 0.26, 1.08)

# Create a data frame
hydrophobicity_df <- data.frame(Amino_Acid = amino_acids, Hydrophobicity = hydrophobicity_values)

hydrophobicity_df

calculate_hydrophobicity <- function(sequence, hydrophobicity_df) {
  sequence <- toupper(sequence)
  # Calculate the hydrophobicity scores for each amino acid, and sum them
  scores <- sapply(strsplit(sequence, '')[[1]], function(aa) {
    match_aa <- match(aa, hydrophobicity_df$Amino_Acid)
    if (!is.na(match_aa)) {
      return(hydrophobicity_df$Hydrophobicity[match_aa])
    } else {
      # If an amino acid is not found, return NA
      return(NA)
    }
  })
  # If there are any NAs in the scores, return NA; otherwise, return the sum
  if (any(is.na(scores))) {
    return(NA)
  } else {
    return(sum(scores))
  }
}

```

# Prep feature df

```{r}

feature_df <- dehashed_combined_tcr_seq %>%
  select(-cdr3_nt) %>%
  rowwise() %>%
  mutate(cdr3_length = nchar(cdr3),
         cdr3_hydrophobicity = calculate_hydrophobicity(cdr3, hydrophobicity_df)) %>%
  ungroup()

feature_df

```

# Plot CDR3 hydrophobicity and length by irAE status

```{r}

custom_labeller <- as_labeller(c("cdr3_length" = "CDR3\nAA length", "cdr3_hydrophobicity" = "CDR3\nhydrophobicity"))

plot_df <- feature_df %>%
  pivot_longer(cols = starts_with("cdr3_"), values_to = "value") %>%
  filter(chain != "Multi")

ggplot(plot_df, aes(x = chain, y = value, fill = irAE)) +
  geom_boxplot() +
  facet_wrap(~ name, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Chain", fill = "irAE", y = "Value")

# Filter the data for statistical test
test_data <- plot_df %>%
  filter(chain == "TRA",
         name == "cdr3_hydrophobicity")

# Calculate the Wilcoxon rank-sum test statistic and p-value
wilcox_result <- wilcox.test(test_data$value ~ test_data$irAE)

# Print the results
print(wilcox_result)

```

# Find CDR3 sequences differentially encountered in yes/no irAE

```{r}

diff_df <- feature_df %>%
  select(barcode, cdr3, chain, Patient_ID) %>%
  filter(cdr3 != "None",
         chain != "Multi") %>%
  mutate(irAE = case_when(
    startsWith(Patient_ID, "A") ~ "No",
    startsWith(Patient_ID, "B") ~ "Yes",
    startsWith(Patient_ID, "M") ~ "Yes")) %>%
  select(-Patient_ID) %>%
  group_by(cdr3, irAE, chain) %>%
  summarize(count = n()) %>%
  ungroup() %>%
  pivot_wider(names_from = c(irAE, chain), values_from = count, values_fill = 0) %>%
  rowwise() %>%
  mutate(tra_diff = abs(No_TRA - Yes_TRA),
         trb_diff = abs(No_TRB - Yes_TRB)) %>%
  ungroup()

diff_df_tra <- diff_df %>%
  select(-No_TRB, -Yes_TRB, -trb_diff) %>%
  filter(tra_diff > 3) %>%
  arrange(-tra_diff)

diff_df_tra

diff_df_trb <- diff_df %>%
  select(-No_TRA, -Yes_TRA, -tra_diff) %>%
  filter(trb_diff > 3) %>%
  arrange(-trb_diff)

diff_df_trb

```

# Test patient bias (TRA, expect same for TRB)

```{r}

# looks like most dramatically differentially encountered CDR3 sequences are due to single patient bias

bias_df_tra <- feature_df %>%
  filter(cdr3 %in% diff_df_tra$cdr3,
         chain == "TRA") %>%
  group_by(Patient_ID, cdr3) %>%
  summarize(count = n()) %>%
  arrange(cdr3, -count) %>%
  filter(cdr3 %in% head(diff_df_tra$cdr3))

bias_df_tra

```

# Plot CDR3 hydrophobicity and length by irAE development for differentially encountered CDR3 sequences

```{r}

custom_labeller <- as_labeller(c("cdr3_length" = "CDR3\nAA length", "cdr3_hydrophobicity" = "CDR3\nhydrophobicity"))

plot_df <- feature_df %>%
  filter(cdr3 %in% diff_df_tra$cdr3 & chain == "TRA" | cdr3 %in% diff_df_trb$cdr3 & chain == "TRB") %>%
  pivot_longer(cols = starts_with("cdr3_"), values_to = "value") %>%
  filter(chain != "Multi")

ggplot(plot_df, aes(x = chain, y = value, fill = irAE)) +
  geom_boxplot() +
  facet_wrap(~ name, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Chain", fill = "irAE", y = "Value")

# Filter the data for statistical test
test_data <- plot_df %>%
  filter(chain == "TRB",
         name == "cdr3_hydrophobicity")

# Calculate the Wilcoxon rank-sum test statistic and p-value
wilcox_result <- wilcox.test(test_data$value ~ test_data$irAE)

# Print the results
print(wilcox_result)

```

# Plot CDR3 hydrophobicity and length by irAE development for differentially encountered CDR3 sequences IGNORING HOW MANY TIMES EACH APPEARS

```{r}

custom_labeller <- as_labeller(c("cdr3_length" = "CDR3\nAA length", "cdr3_hydrophobicity" = "CDR3\nhydrophobicity"))

plot_df <- feature_df %>%
  filter(cdr3 %in% diff_df_tra$cdr3 & chain == "TRA" | cdr3 %in% diff_df_trb$cdr3 & chain == "TRB") %>%
  group_by(cdr3, chain, irAE) %>%
  slice(1) %>%
  ungroup() %>%
  pivot_longer(cols = starts_with("cdr3_"), values_to = "value") %>%
  filter(chain != "Multi")

ggplot(plot_df, aes(x = chain, y = value, fill = irAE)) +
  geom_boxplot() +
  facet_wrap(~ name, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Chain", fill = "irAE", y = "Value")

# Filter the data for statistical test
test_data <- plot_df %>%
  filter(chain == "TRB",
         name == "cdr3_length")

# Calculate the Wilcoxon rank-sum test statistic and p-value
wilcox_result <- wilcox.test(test_data$value ~ test_data$irAE)

# Print the results
print(wilcox_result)

```

# Plot pgen for differentially encountered CDR3 sequences

```{r}

plot_df <- igor_combined %>%
  filter(cdr3 %in% diff_df_tra$cdr3 & chain == "TRA" | cdr3 %in% diff_df_trb$cdr3 & chain == "TRB")

ggplot(plot_df, aes(x = chain, y = -log10(pgen), fill = irAE)) +
  geom_boxplot() +
  labs(x = "Chain", y = "-log10(pgen)")

```

# Plot pgen for differentially encountered CDR3 sequences IGNORING HOW MANY TIMES EACH APPEARS

```{r}

plot_df <- igor_combined %>%
  filter(cdr3 %in% diff_df_tra$cdr3 & chain == "TRA" | cdr3 %in% diff_df_trb$cdr3 & chain == "TRB") %>%
  group_by(cdr3, chain, patient_group) %>%
  slice(1) %>%
  ungroup()

ggplot(plot_df, aes(x = chain, y = -log10(pgen), fill = irAE)) +
  geom_boxplot() +
  labs(x = "Chain", y = "-log10(pgen)")

# Filter the data for statistical test
test_data <- plot_df %>%
  filter(chain == "TRB")

# Calculate the Wilcoxon rank-sum test statistic and p-value
wilcox_result <- wilcox.test(test_data$pgen ~ test_data$irAE)

# Print the results
print(wilcox_result)

```

# Look at public TCRs only

```{r}

alpha_public_df <- feature_df %>%
  select(barcode, cdr3, chain, Patient_ID) %>%
  filter(cdr3 != "None",
         chain == "TRA") %>%
  group_by(cdr3, Patient_ID) %>%
  summarize(count = n()) %>%
  ungroup() %>%
  pivot_wider(names_from = c(Patient_ID), values_from = count, values_fill = 0) %>%
  rowwise() %>%
  filter(sum(c_across(-cdr3)) > 1)

beta_public_df <- feature_df %>%
  select(barcode, cdr3, chain, Patient_ID) %>%
  filter(cdr3 != "None",
         chain == "TRB") %>%
  group_by(cdr3, Patient_ID) %>%
  summarize(count = n()) %>%
  ungroup() %>%
  pivot_wider(names_from = c(Patient_ID), values_from = count, values_fill = 0) %>%
  rowwise() %>%
  filter(sum(c_across(-cdr3)) > 1)

public_feature_df <- feature_df %>%
  filter(cdr3 %in% alpha_public_df$cdr3 & chain == "TRA" | cdr3 %in% beta_public_df$cdr3 & chain == "TRB")

public_igor_df <- igor_combined %>%
  filter(cdr3 %in% alpha_public_df$cdr3 & chain == "TRA" | cdr3 %in% beta_public_df$cdr3 & chain == "TRB")

```

# Plot CDR3 hydrophobicity and length by irAE status for public TCRs

```{r}

custom_labeller <- as_labeller(c("cdr3_length" = "CDR3\nAA length", "cdr3_hydrophobicity" = "CDR3\nhydrophobicity"))

plot_df <- public_feature_df %>%
  pivot_longer(cols = starts_with("cdr3_"), values_to = "value") %>%
  filter(chain != "Multi")

ggplot(plot_df, aes(x = chain, y = value, fill = irAE)) +
  geom_boxplot() +
  facet_wrap(~ name, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Chain", fill = "irAE", y = "Value")

```

# Plot pgen scores by irAE development for public TCRs

```{r}

ggplot(public_igor_df, aes(x = chain, y = -log10(pgen), fill = irAE)) +
  geom_boxplot() +
  labs(x = "Chain", y = "-log10(pgen)")

```

# Stratify analysis by expansion degree

```{r}

feature_df <- feature_df %>%
  group_by(Patient_ID, cdr3) %>%
  mutate(count = n()) %>%
  ungroup() %>%
  mutate(expansion_degree = case_when(
    count == 1 ~ "1",
    count > 1 & count <= 5 ~ "1-5",
    count > 5 & count <= 20 ~ "5-20",
    count > 20 & count <= 100 ~ "20-100",
    count > 100 ~ "> 100"))

feature_df$expansion_degree <- factor(feature_df$expansion_degree, levels = c("1", "1-5", "5-20", "20-100", "> 100"))

plot_df <- feature_df %>%
  filter(chain != "Multi")

ggplot(plot_df, aes(x = chain, y = cdr3_length, fill = irAE)) +
  geom_boxplot() +
  facet_wrap(~expansion_degree, scales = "free_y") +
  labs(x = "Chain", fill = "irAE", y = "CDR3 length (AA)")

ggplot(plot_df, aes(x = chain, y = cdr3_hydrophobicity, fill = irAE)) +
  geom_boxplot() +
  facet_wrap(~expansion_degree, scales = "free_y") +
  labs(x = "Chain", fill = "irAE", y = "CDR3 hydrophobicity")

igor_combined <- igor_combined %>%
  group_by(Patient_ID, cdr3) %>%
  mutate(count = n()) %>%
  ungroup() %>%
  mutate(expansion_degree = case_when(
    count == 1 ~ "1",
    count > 1 & count <= 5 ~ "1-5",
    count > 5 & count <= 20 ~ "5-20",
    count > 20 & count <= 100 ~ "20-100",
    count > 100 ~ "> 100"))

igor_combined$expansion_degree <- factor(igor_combined$expansion_degree, levels = c("1", "1-5", "5-20", "20-100", "> 100"))

ggplot(igor_combined, aes(x = chain, y = -log10(pgen), fill = irAE)) +
  geom_boxplot() +
  facet_wrap(~expansion_degree) +
  labs(x = "Chain", y = "-log10(pgen)")

```