---
title: "Identification of Pathogenic Immune Cell Subsets Associated with Checkpoint Inhibitor-induced Myocarditis"
output: pdf_document
---

 - data reproduction of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9397491/
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(Seurat)
library(Matrix)
library(ggplot2)
library(dplyr)
library(tidyr)
library(stringr)
library(circlize)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Read in non-control TCR-seq data

```{r}

data_directory <- "../raw_data"

# Identify the pattern to extract patient IDs using regular expressions
pattern <- "GSM\\d+_([A-Z\\d\\-]+)-filtered_contig_annotations"

# Get a list of CSV files in the directory
csv_files <- list.files(data_directory, pattern = ".csv", full.names = TRUE)

# Initialize an empty list to store data frames
data_frames <- list()

# Loop through the list of CSV files
for (file in csv_files) {
  # Extract patient ID from the file name using regex
  patient_id <- gsub("GSM(\\d+)_(.*?)\\-filtered_contig_annotations.csv", "\\2", basename(file))
  
  # Read the CSV file into a data frame
  df <- read.csv(file)
  
  # Add a column for patient ID
  df$Patient_ID <- patient_id
  
  # Append the data frame to the list
  data_frames[[patient_id]] <- df
  
}

# Combine all data frames into a single data frame
combined_tcr_seq <- bind_rows(data_frames)

```

# Read in dehashed RDS

```{r}

dehashed <- readRDS("../saved_data/dehashed.rds")

```

# Dehash TCRseq data, lose some barcodes in process (0-lets, multiplets, no hash marker)

```{r}

A6 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$A6)) %>%
  mutate(Patient_ID = "A6")

A9 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$A9)) %>%
  mutate(Patient_ID = "A9")

B13 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$B13)) %>%
  mutate(Patient_ID = "B13")

B9 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A6-A9-B13-B9"),
         barcode %in% colnames(dehashed$B9)) %>%
  mutate(Patient_ID = "B9")

A8 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A8-B11"),
         barcode %in% colnames(dehashed$A8)) %>%
  mutate(Patient_ID = "A8")

B11 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A8-B11"),
         barcode %in% colnames(dehashed$B11)) %>%
  mutate(Patient_ID = "B11")

B12 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("B12-MCE6-MCE7"),
         barcode %in% colnames(dehashed$B12)) %>%
  mutate(Patient_ID = "B12")

MCE6 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("B12-MCE6-MCE7"),
         barcode %in% colnames(dehashed$MCE6)) %>%
  mutate(Patient_ID = "MCE6")

MCE7 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("B12-MCE6-MCE7"),
         barcode %in% colnames(dehashed$MCE7)) %>%
  mutate(Patient_ID = "MCE7")

A10 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$A10)) %>%
  mutate(Patient_ID = "A10")

B10 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$B10)) %>%
  mutate(Patient_ID = "B10")

MCE12 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$MCE12)) %>%
  mutate(Patient_ID = "MCE12")

MCE13 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A10-B10-MCE12-MCE13"),
         barcode %in% colnames(dehashed$MCE13)) %>%
  mutate(Patient_ID = "MCE13")

A5 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A5-B14"),
         barcode %in% colnames(dehashed$A5)) %>%
  mutate(Patient_ID = "A5")

B14 <- combined_tcr_seq %>%
  filter(Patient_ID %in% c("A5-B14"),
         barcode %in% colnames(dehashed$B14)) %>%
  mutate(Patient_ID = "B14")

dehashed_combined_tcr_seq <- combined_tcr_seq %>%
  filter(!(Patient_ID %in% c("A5-B14", "A10-B10-MCE12-MCE13", "B12-MCE6-MCE7", "A6-A9-B13-B9", "A8-B11")),
         productive != "False")

# Define a list of data frames for each patient or barcode
patient_dataframes <- list(A6, A9, B13, B9, A8, B11, B12, MCE6, MCE7, A10, B10, MCE12, MCE13, A5, B14)

# Concatenate the data frames into a single data frame
dehashed_combined_tcr_seq <- bind_rows(dehashed_combined_tcr_seq, patient_dataframes)

dehashed_combined_tcr_seq <- dehashed_combined_tcr_seq %>%
  mutate(irAE = case_when(
    startsWith(Patient_ID, "A") ~ "No",
    startsWith(Patient_ID, "B") ~ "Yes",
    startsWith(Patient_ID, "M") ~ "Yes"))

dehashed_combined_tcr_seq

write.csv(dehashed_combined_tcr_seq, "../saved_data/dehashed_tcr_seq.csv")

```

# Read in final RNA T cell Seurat with TCR metadata

```{r}

tcr_rna <- readRDS("../saved_data/final_seurat_T_cells.rds")

```

# Tibble metadata, add rownames

```{r}

dehashed_combined_tcr_seq_with_rna <- as_tibble(tcr_rna@meta.data)

dehashed_combined_tcr_seq_with_rna$TCR_name <- colnames(tcr_rna)

dehashed_combined_tcr_seq_with_rna

```

# Prep TRA stitchr df

```{r}

bad_alpha <- dehashed_combined_tcr_seq_with_rna %>%
  rename(TRAV = v_gene_TRA, TRAJ = j_gene_TRA, TRBV = v_gene_TRB, TRBJ = j_gene_TRB, TRAC = c_gene_TRA, TRBC = c_gene_TRB, TRA_CDR3 = cdr3_TRA, TRB_CDR3 = cdr3_TRB) %>%
  select(TCR_name, TRAV, TRAJ, TRA_CDR3, TRBV, TRBJ, TRB_CDR3, TRAC, TRBC) %>%
  filter_at(vars(starts_with("TRA")), any_vars(. == "None"))

stitchr_alpha_data <- dehashed_combined_tcr_seq_with_rna %>%
  rename(TRAV = v_gene_TRA, TRAJ = j_gene_TRA, TRBV = v_gene_TRB, TRBJ = j_gene_TRB, TRAC = c_gene_TRA, TRBC = c_gene_TRB, TRA_CDR3 = cdr3_TRA, TRB_CDR3 = cdr3_TRB) %>%
  select(TCR_name, TRAV, TRAJ, TRA_CDR3, TRBV, TRBJ, TRB_CDR3, TRAC, TRBC) %>%
  filter(!(TCR_name %in% bad_alpha$TCR_name)) %>%
  mutate(TRA_leader	= NA,
         TRB_leader = NA,
         Linker = NA,
         Link_order = NA,
         TRA_5_prime_seq = NA,
         TRA_3_prime_seq = NA,
         TRB_5_prime_seq = NA,
         TRB_3_prime_seq = NA)

stitchr_alpha_data

# 30,279 rows

write.table(stitchr_alpha_data, file = "../saved_data/tcrseq_tra_for_stitchr.tsv", sep = "\t", quote = FALSE, row.names = FALSE, na = "")

bad_beta <- dehashed_combined_tcr_seq_with_rna %>%
    rename(TRAV = v_gene_TRA, TRAJ = j_gene_TRA, TRBV = v_gene_TRB, TRBJ = j_gene_TRB, TRAC = c_gene_TRA, TRBC = c_gene_TRB, TRA_CDR3 = cdr3_TRA, TRB_CDR3 = cdr3_TRB) %>%
  select(TCR_name, TRAV, TRAJ, TRA_CDR3, TRBV, TRBJ, TRB_CDR3, TRAC, TRBC) %>%
  filter_at(vars(starts_with("TRB")), any_vars(. == "None"))

stitchr_beta_data <- dehashed_combined_tcr_seq_with_rna %>%
  rename(TRAV = v_gene_TRA, TRAJ = j_gene_TRA, TRBV = v_gene_TRB, TRBJ = j_gene_TRB, TRAC = c_gene_TRA, TRBC = c_gene_TRB, TRA_CDR3 = cdr3_TRA, TRB_CDR3 = cdr3_TRB) %>%
  select(TCR_name, TRAV, TRAJ, TRA_CDR3, TRBV, TRBJ, TRB_CDR3, TRAC, TRBC) %>%
  filter(!(TCR_name %in% bad_beta$TCR_name)) %>%
  mutate(TRA_leader	= NA,
         TRB_leader = NA,
         Linker = NA,
         Link_order = NA,
         TRA_5_prime_seq = NA,
         TRA_3_prime_seq = NA,
         TRB_5_prime_seq = NA,
         TRB_3_prime_seq = NA)

stitchr_beta_data

# 30,313 rows

write.table(stitchr_beta_data, file = "../saved_data/tcrseq_trb_for_stitchr.tsv", sep = "\t", quote = FALSE, row.names = FALSE, na = "")

```

# Read in stitchr output

```{r}

stitchr_alpha_output <- read.delim("../saved_data/stitchr_tra_out.tsv", header = TRUE, sep = "\t")

stitchr_alpha_output

stitchr_beta_output <- read.delim("../saved_data/stitchr_trb_out.tsv", header = TRUE, sep = "\t")

stitchr_beta_output

stitchr_output <- stitchr_alpha_output %>%
  full_join(stitchr_beta_output, by = colnames(stitchr_alpha_output))

stitchr_output

```

# Prep IGoR input

```{r}

stitchr_output_alpha_df <- stitchr_output %>%
  filter(!(is.na(TRA_nt)),
         nchar(TRA_nt) > 1)

stitchr_output_alpha_df
# 30,277 rows

write.csv(stitchr_output_alpha_df, file = "../saved_data/igor_alpha_input.csv", row.names = FALSE)

stitchr_output_beta_df <- stitchr_output %>%
  filter(!(is.na(TRB_nt)),
         nchar(TRB_nt) > 1)

stitchr_output_beta_df
# 30,263 rows

write.csv(stitchr_output_beta_df, file = "../saved_data/igor_beta_input.csv", row.names = FALSE)

```

# Read in IGoR output

```{r}

igor_alpha <- read.csv("../saved_data/igor_output_alpha.csv") %>%
  rename(barcode = TCR_name) %>%
  left_join(dehashed_combined_tcr_seq, by = "barcode") %>%
  mutate(patient_group = case_when(
    startsWith(Patient_ID, "A") ~ "ici_no_irae",
    startsWith(Patient_ID, "B") ~ "ici_non_myocarditis_irae",
    startsWith(Patient_ID, "M") ~ "ici_myocarditis_irae")) %>%
  mutate(irAE = case_when(
    startsWith(Patient_ID, "A") ~ "No",
    startsWith(Patient_ID, "B") ~ "Yes",
    startsWith(Patient_ID, "M") ~ "Yes")) %>%
  select(patient_group, irAE, chain, pgen, barcode, cdr3) %>%
  filter(chain == "TRA",
         pgen != "NaN")

igor_combined <- read.csv("../saved_data/igor_output_beta.csv") %>%
  rename(barcode = TCR_name) %>%
  left_join(dehashed_combined_tcr_seq, by = "barcode") %>%
  mutate(patient_group = case_when(
    startsWith(Patient_ID, "A") ~ "ici_no_irae",
    startsWith(Patient_ID, "B") ~ "ici_non_myocarditis_irae",
    startsWith(Patient_ID, "M") ~ "ici_myocarditis_irae")) %>%
  mutate(irAE = case_when(
    startsWith(Patient_ID, "A") ~ "No",
    startsWith(Patient_ID, "B") ~ "Yes",
    startsWith(Patient_ID, "M") ~ "Yes")) %>%
  select(Patient_ID, patient_group, irAE, chain, pgen, barcode, cdr3) %>%
  filter(chain == "TRB",
         pgen != "NaN") %>%
  full_join(igor_alpha, by = c("barcode", "chain", "patient_group", "pgen", "cdr3", "irAE"))

igor_combined

```

# Prep IMGT/HighV-Quest input

```{r}

fasta_df_alpha <- stitchr_output %>%
  filter(!(is.na(TRA_nt)),
         nchar(TRA_nt) > 1)

fasta_df_alpha

# Function to convert a data frame to FASTA format
df_to_fasta_alpha <- function(df, output_file) {
  # Open/create the output file for writing
  fasta_file <- file(output_file, "w")
  
  # Iterate through rows of the data frame
  for (i in 1:nrow(df)) {
    # Combine name columns and add a suffix numbering
    label <- df$TCR_name[i]

    # Write the sequence and label to the FASTA file
    writeLines(paste(">", label), con = fasta_file)
    writeLines(df$TRA_nt[i], con = fasta_file)
  }
  
  # Close the output file
  close(fasta_file)
}

# Call the function to convert the data frame to FASTA format
df_to_fasta_alpha(fasta_df_alpha, "../saved_data/vdjseq_alpha.fasta")

fasta_df_beta <- stitchr_output %>%
  filter(!(is.na(TRA_nt)),
         nchar(TRB_nt) > 1)

fasta_df_beta

# Function to convert a data frame to FASTA format
df_to_fasta_beta <- function(df, output_file) {
  # Open/create the output file for writing
  fasta_file <- file(output_file, "w")
  
  # Iterate through rows of the data frame
  for (i in 1:nrow(df)) {
    # Combine name columns and add a suffix numbering
    label <- df$TCR_name[i]

    # Write the sequence and label to the FASTA file
    writeLines(paste(">", label), con = fasta_file)
    writeLines(df$TRB_nt[i], con = fasta_file)
  }
  
  # Close the output file
  close(fasta_file)
}

# Call the function to convert the data frame to FASTA format
df_to_fasta_beta(fasta_df_beta, "../saved_data/vdjseq_beta.fasta")

```

# Analyze IMGT/HighV-Quest output 1: summary

```{r}

imgt_df_1_alpha <- read.csv("../saved_data/myocarditis_fixed_tra/1_Summary.csv") %>%
  mutate(J_allele = str_extract(J.GENE.and.allele, "\\*([0-9]{2})")) %>%
  group_by(J_allele) %>%
  summarize(count = n())
  #group_by(V.DOMAIN.Functionality) %>%
  #summarize(count = n())

imgt_df_1_alpha
# 100% productive, V alleles all *01, J alleles mostly *01 but some *02/*03

imgt_df_1_beta <- read.csv("../saved_data/myocarditis_fixed_trb/1_Summary.csv") %>%
  mutate(J_allele = str_extract(J.GENE.and.allele, "\\*([0-9]{2})")) %>%
  group_by(J_allele) %>%
  summarize(count = n())
  #group_by(V.DOMAIN.Functionality) %>%
  #summarize(count = n())

imgt_df_1_beta
# 100% productive, V and J alleles all *01

```

# Analyze IMGT/HighV-Quest output 3: CDR1/2 and N region lengths

```{r}

imgt_df_3_alpha <- read.csv("../saved_data/myocarditis_fixed_tra/3_Nt-sequences.csv") %>%
  select(Sequence.ID, N.REGION, N1.REGION, N2.REGION, N3.REGION) %>%
  rename(barcode = Sequence.ID,
         N.REGION_TRA = N.REGION,
         N1.REGION_TRA = N1.REGION,
         N2.REGION_TRA = N2.REGION,
         N3.REGION_TRA = N3.REGION)

imgt_df_3_alpha

imgt_df_3 <- read.csv("../saved_data/myocarditis_fixed_trb/3_Nt-sequences.csv") %>%
  select(Sequence.ID, N.REGION, N1.REGION, N2.REGION, N3.REGION, CDR1.IMGT, CDR2.IMGT) %>%
  rename(barcode = Sequence.ID,
         N.REGION_TRB = N.REGION,
         N1.REGION_TRB = N1.REGION,
         N2.REGION_TRB = N2.REGION,
         N3.REGION_TRB = N3.REGION,
         cdr1 = as.character(translate(DNAString(DR1.IMGT))),
         cdr2 = as.character(translate(DNAString(CDR2.IMGT)))) %>%
  full_join(imgt_df_3_alpha, by = "barcode")

imgt_df_3

write.csv(imgt_df_3, "../saved_data/imgt_cdr1_2_n_regions.csv")

```

# Translate CDR1/2

```{r}

# Install and load the Biostrings package
#if (!requireNamespace("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#BiocManager::install("Biostrings")
#library(Biostrings)

# Example DNA sequence
dna_sequence <- DNAString("ATG")

x <- as.character(translate(dna_sequence))

x

```