---
title: "T-cell invigoration to tumour burden ratio associated with anti-PD-1 response"
output: html_notebook
---

 - data reproduction of https://www.nature.com/articles/nature22079

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(TAR)
library(Seurat)
library(Matrix)
library(patchwork)
library(ggplot2)
library(data.table)

```
# Try to import data

```{r}

# Load gene expression data
raw_counts <- fread("GSE96578_AH_PBMC_RNAseq_rawcounts.txt.gz")

# Load cell barcode data
cell_barcodes <- fread("matrix.txt.gz", header = FALSE, sep = "\t")

# Load sample information (family.soft.gz and family.xml.tgz)
# You may need to extract the necessary information from these files.
# This will depend on the specific structure of these files.

# For example, if family.soft.gz contains sample metadata in SOFT format:
# family_info <- read.soft("family.soft.gz")

# If family.xml.tgz contains metadata in XML format:
# Extract and read XML data.

# Combine cell barcode and gene expression data
gene_expression_matrix <- t(as.matrix(raw_counts))
colnames(gene_expression_matrix) <- cell_barcodes$V1

# Create a Seurat object
seurat_obj <- CreateSeuratObject(counts = gene_expression_matrix)

```