---
title: "T cell characteristics associated with irAE from ICI in melanoma"
output: pdf_document
---

 - data reproduction of https://www.nature.com/articles/s41591-021-01623-z

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(Seurat)
library(patchwork)
library(ggplot2)
library(dplyr)
library(tidyr)
library(tidyverse)
library(forcats)
library(vegan)
library(readxl)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Read in bulk ImmunoSEQ data

```{r}

# List of file names
file_names <- c(
  "../raw_data/ImmunoSEQ/WUMENLO_pre.tsv",
  "../raw_data/ImmunoSEQ/WUMENLO_on.tsv",
  "../raw_data/ImmunoSEQ/YUAMIGO_on.tsv",
  "../raw_data/ImmunoSEQ/YUCADA_on.tsv",
  "../raw_data/ImmunoSEQ/YUFROZE_on.tsv",
  "../raw_data/ImmunoSEQ/YUHONEY_on.tsv",
  "../raw_data/ImmunoSEQ/YUNANCY_on.tsv",
  "../raw_data/ImmunoSEQ/YUPOGER_on.tsv",
  "../raw_data/ImmunoSEQ/YUTACI_on.tsv",
  "../raw_data/ImmunoSEQ/YUAMIGO_pre.tsv",
  "../raw_data/ImmunoSEQ/YUCADA_pre.tsv",
  "../raw_data/ImmunoSEQ/YUFROZE_pre.tsv",
  "../raw_data/ImmunoSEQ/YUHONEY_pre.tsv",
  "../raw_data/ImmunoSEQ/YUNANCY_pre.tsv",
  "../raw_data/ImmunoSEQ/YUPOGER_pre.tsv",
  "../raw_data/ImmunoSEQ/YUTACI_pre.tsv",
  "../raw_data/ImmunoSEQ/YUALOE_on.tsv",
  "../raw_data/ImmunoSEQ/YUBROMO_on.tsv",
  "../raw_data/ImmunoSEQ/YUDEBIT_on.tsv",
  "../raw_data/ImmunoSEQ/YUHERN_on.tsv",
  "../raw_data/ImmunoSEQ/YULAMB_on.tsv",
  "../raw_data/ImmunoSEQ/YUPESO_on.tsv",
  "../raw_data/ImmunoSEQ/YUQIP_on.tsv",
  "../raw_data/ImmunoSEQ/YUALOE_pre.tsv",
  "../raw_data/ImmunoSEQ/YUBROMO_pre.tsv",
  "../raw_data/ImmunoSEQ/YUDEBIT_pre.tsv",
  "../raw_data/ImmunoSEQ/YUHERN_pre.tsv",
  "../raw_data/ImmunoSEQ/YULAMB_pre.tsv",
  "../raw_data/ImmunoSEQ/YUPESO_pre.tsv",
  "../raw_data/ImmunoSEQ/YUQIP_pre.tsv"
)

# Create an empty data frame to store the combined data
combined_immunoseq_data <- data.frame()

# Process each file and append it to the combined_data data frame
for (file_name in file_names) {
  # Read the TSV file
  data <- read_tsv(file_name)
  
  extracted_ici_status <- gsub(".*/[^/]+_(on|pre)\\.tsv$", "\\1", file_name)
  extracted_patient_id <- gsub(".*/([^/]+)_(pre|on)\\.tsv$", "\\1", file_name)
  
  data$ici_status <- extracted_ici_status
  data$patient_id <- extracted_patient_id
  
  # Append the data to the combined_data data frame
  combined_immunoseq_data <- bind_rows(combined_immunoseq_data, data)
}

```

# Read in metadata, clean, join with diversity scores

```{r}

irae_statuses <- read_excel("../raw_data/supp_tables.xlsx", sheet = "cleaned_table_s1")

irae_statuses <- irae_statuses %>%
  select(`Patient ID`, `Highest irAE grade`) %>%
  rename(patient_id = `Patient ID`, severe_irae = `Highest irAE grade`) %>%
  mutate(severe_irae = if_else(severe_irae %in% c(3, 4), "Yes", "No"))

```

# Join irAE development with immunoseq data

```{r}

combined_immunoseq_data <- combined_immunoseq_data %>%
  left_join(irae_statuses, by = "patient_id")

combined_immunoseq_data

```

# Count repeats of various TCR measures, calculate diversity scores, join with metadata

```{r}

rearrangement_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(rearrangement))) %>%
  group_by(rearrangement, patient_id, ici_status) %>%
  mutate(rearrangement_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(rearrangement_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(rearrangement_count))) %>%
  mutate(rearrangement_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(rearrangement_score_delta = rearrangement_score[ici_status == "on"] - rearrangement_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

amino_acid_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(amino_acid))) %>%
  group_by(amino_acid, patient_id, ici_status) %>%
  mutate(amino_acid_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(amino_acid_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(amino_acid_count))) %>%
  mutate(amino_acid_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(amino_acid_score_delta = amino_acid_score[ici_status == "on"] - amino_acid_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

rearrangement_trunc_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(rearrangement_trunc))) %>%
  group_by(rearrangement_trunc, patient_id, ici_status) %>%
  mutate(rearrangement_trunc_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(rearrangement_trunc_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(rearrangement_trunc_count))) %>%
  mutate(rearrangement_trunc_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(rearrangement_trunc_score_delta = rearrangement_trunc_score[ici_status == "on"] - rearrangement_trunc_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

extended_rearrangement_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(extended_rearrangement))) %>%
  group_by(extended_rearrangement, patient_id, ici_status) %>%
  mutate(extended_rearrangement_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(extended_rearrangement_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(extended_rearrangement_count))) %>%
  mutate(extended_rearrangement_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(extended_rearrangement_score_delta = extended_rearrangement_score[ici_status == "on"] - extended_rearrangement_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

cdr1_rearrangement_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(cdr1_rearrangement))) %>%
  group_by(cdr1_rearrangement, patient_id, ici_status) %>%
  mutate(cdr1_rearrangement_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(cdr1_rearrangement_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(cdr1_rearrangement_count))) %>%
  mutate(cdr1_rearrangement_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(cdr1_rearrangement_score_delta = cdr1_rearrangement_score[ici_status == "on"] - cdr1_rearrangement_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

cdr1_amino_acid_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(cdr1_amino_acid))) %>%
  group_by(cdr1_amino_acid, patient_id, ici_status) %>%
  mutate(cdr1_amino_acid_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(cdr1_amino_acid_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(cdr1_amino_acid_count))) %>%
  mutate(cdr1_amino_acid_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(cdr1_amino_acid_score_delta = cdr1_amino_acid_score[ici_status == "on"] - cdr1_amino_acid_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

cdr2_rearrangement_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(cdr2_rearrangement))) %>%
  group_by(cdr2_rearrangement, patient_id, ici_status) %>%
  mutate(cdr2_rearrangement_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(cdr2_rearrangement_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(cdr2_rearrangement_count))) %>%
  mutate(cdr2_rearrangement_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(cdr2_rearrangement_score_delta = cdr2_rearrangement_score[ici_status == "on"] - cdr2_rearrangement_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

cdr2_amino_acid_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(cdr2_amino_acid))) %>%
  group_by(cdr2_amino_acid, patient_id, ici_status) %>%
  mutate(cdr2_amino_acid_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(cdr2_amino_acid_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(cdr2_amino_acid_count))) %>%
  mutate(cdr2_amino_acid_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(cdr2_amino_acid_score_delta = cdr2_amino_acid_score[ici_status == "on"] - cdr2_amino_acid_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

cdr3_rearrangement_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(cdr3_rearrangement))) %>%
  group_by(cdr3_rearrangement, patient_id, ici_status) %>%
  mutate(cdr3_rearrangement_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(cdr3_rearrangement_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(cdr3_rearrangement_count))) %>%
  mutate(cdr3_rearrangement_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(cdr3_rearrangement_score_delta = cdr3_rearrangement_score[ici_status == "on"] - cdr3_rearrangement_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

cdr3_amino_acid_diversity_df <- combined_immunoseq_data %>%
  filter(!(is.na(cdr3_amino_acid))) %>%
  group_by(cdr3_amino_acid, patient_id, ici_status) %>%
  mutate(cdr3_amino_acid_count = n()) %>%
  ungroup() %>%
  group_by(patient_id, ici_status) %>%
  # Calculate Shannon diversity index (H)
  mutate(shannon = diversity(cdr3_amino_acid_count)) %>%
  # Calculate Pielou's evenness index (J)
  mutate(pielou_evenness = shannon / log(specnumber(cdr3_amino_acid_count))) %>%
  mutate(cdr3_amino_acid_score = 1 - pielou_evenness) %>%
  slice(1) %>%
  ungroup() %>%
  group_by(patient_id) %>%
  summarize(cdr3_amino_acid_score_delta = cdr3_amino_acid_score[ici_status == "on"] - cdr3_amino_acid_score[ici_status == "pre"]) %>%
  left_join(irae_statuses, by = "patient_id")

```

# Join dfs to plot all at once

```{r}

joined_diversity_df <- rearrangement_diversity_df %>%
  full_join(amino_acid_diversity_df) %>%
  full_join(rearrangement_trunc_diversity_df) %>%
  full_join(extended_rearrangement_diversity_df) %>%
  full_join(cdr1_rearrangement_diversity_df) %>%
  full_join(cdr1_amino_acid_diversity_df) %>%
  full_join(cdr2_rearrangement_diversity_df) %>%
  full_join(cdr3_amino_acid_diversity_df) %>%
  full_join(cdr3_rearrangement_diversity_df) %>%
  full_join(cdr3_amino_acid_diversity_df) %>%
  pivot_longer(cols = -c(patient_id, severe_irae), 
               names_to = "variable",
               values_to = "score_delta") %>%
  mutate(origin = sub("_score_delta", "", variable)) %>%
  select(-variable)

joined_diversity_df

```

# Plot TCR diversity score changes upon treatment by future irAE status

```{r}

# Create a custom label function
custom_label_function <- function(variable, value) {
  # Define custom labels based on the values of 'origin'
  custom_labels <- c(
    extended_rearrangement = "Extended\n rearrangement",
    rearrangement_trunc = "Truncated\nrearrangement",
    amino_acid = "Amino\nacid",
    cdr1_amino_acid = "CDR1\namino acid",
    cdr1_rearrangement = "CDR1\nrearrangement",
    cdr2_amino_acid = "CDR2\namino acid",
    cdr2_rearrangement = "CDR2\nrearrangement",
    cdr3_amino_acid = "CDR3\namino acid",
    cdr3_rearrangement = "CDR3\nrearrangement")
  
  # Return the custom labels
  return(custom_labels[value])
}

ggplot(joined_diversity_df, aes(x = severe_irae, y = score_delta, fill = severe_irae)) +
  geom_boxplot() +
  geom_point(size = 1) +
  facet_wrap(~origin, scales = "free_y", labeller = labeller(origin = custom_label_function)) +
  labs(x = "Severe irAE", y = "Change in clonality\nfrom baseline") +
  scale_fill_manual(values = c("blue", "red")) +
  theme(legend.position = "none",
        strip.text = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.y = element_line(size = 0.5))

ggsave("../figures/complement_of_clonal_evenness.svg")

# the complement of clonal evenness (1-Pielou’s index) is often used to get a clonality score of 0 representing a maximally diverse population with even frequencies and values close to 1, a repertoire driven by clonal dominance. 

```

# Prepare FASTA/Q files for IMGT/HighV-Quest

```{r}

fasta_fastq_prep <- combined_immunoseq_data %>%
  select(extended_rearrangement, ici_status, patient_id) %>%
  filter(!(is.na(extended_rearrangement))) %>%
  group_by(ici_status, patient_id) %>%
  mutate(suffix = row_number()) %>%
  ungroup()

# Function to convert a data frame to FASTA format
df_to_fasta <- function(df, output_file) {
  # Open/create the output file for writing
  fasta_file <- file(output_file, "w")
  
  # Iterate through rows of the data frame
  for (i in 1:nrow(df)) {
    # Combine name columns and add a suffix numbering
    label <- paste(df$ici_status[i], df$patient_id[i], df$suffix[i], sep="_")
    
    # Write the sequence and label to the FASTA file
    writeLines(paste(">", label), con = fasta_file)
    writeLines(df$extended_rearrangement[i], con = fasta_file)
  }
  
  # Close the output file
  close(fasta_file)
}

# Call the function to convert the data frame to FASTA format
df_to_fasta(fasta_fastq_prep, "../saved_data/immunoseq.fasta")

```


# Investigate TCR length and germline-ness as a function of treatment status, box plots

```{r}

ici_length_and_germlineness_df <- combined_immunoseq_data %>%
  select(n1_insertions, n2_insertions, cdr1_rearrangement_length, cdr2_rearrangement_length, cdr3_rearrangement_length, cdr3_length, ici_status) %>%
  gather(key = "Category", value = "Value", -ici_status)

ici_length_and_germlineness_df

custom_labeller <- as_labeller(c("n1_insertions" = "n1\ninsertions", "n2_insertions" = "n2\ninsertions", "cdr1_rearrangement_length" = "cdr1\nrearrangement\nlength (nt)", "cdr2_rearrangement_length" = "cdr2\nrearrangement\nlength (nt)", "cdr3_rearrangement_length" = "cdr3\nrearrangement\nlength (nt)", "cdr3_length" = "cdr3\nlength (nt)"))

# Create the box plot with facets
ggplot(ici_length_and_germlineness_df, aes(x = ici_status, y = Value)) +
  geom_boxplot() +
  facet_wrap(~ Category, scales = "free_y", labeller = custom_labeller) +
  labs(x = "ici_status", y = "Value")

```

# Histograms instead of boxplots, look at CDR1/2 rearrangement lengths

```{r}

# Create the overlaid histograms with facets and colored by ici_status
ggplot(ici_length_and_germlineness_df %>% filter(Category %in% c("cdr1_rearrangement_length", "cdr2_rearrangement_length")), aes(x = Value, fill = ici_status)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(14, 22) +
  scale_fill_manual(values = c("pre" = "blue", "on" = "red"))

```

# Histograms instead of boxplots, look at CDR3 (rearrangement) lengths

```{r}

# Create the overlaid histograms with facets and colored by ici_status
ggplot(ici_length_and_germlineness_df %>% filter(Category %in% c("cdr3_rearrangement_length", "cdr3_length")), aes(x = Value, fill = ici_status)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(0, 80) +
  scale_fill_manual(values = c("pre" = "blue", "on" = "red"))

```

# Histograms instead of boxplots, look at n1/2 insertions

```{r}

# Create the overlaid histograms with facets and colored by ici_status
ggplot(ici_length_and_germlineness_df %>% filter(Category %in% c("n1_insertions", "n2_insertions")), aes(x = Value, fill = ici_status)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(0, 20) +
  scale_fill_manual(values = c("pre" = "blue", "on" = "red"))

```

# Investigate TCR length and germline-ness as a function of irAE development, box plots

```{r}

irae_length_and_germlineness_df <- combined_immunoseq_data %>%
  select(n1_insertions, n2_insertions, cdr1_rearrangement_length, cdr2_rearrangement_length, cdr3_rearrangement_length, cdr3_length, severe_irae) %>%
  gather(key = "Category", value = "Value", -severe_irae)

irae_length_and_germlineness_df

custom_labeller <- as_labeller(c("n1_insertions" = "n1\ninsertions", "n2_insertions" = "n2\ninsertions", "cdr1_rearrangement_length" = "cdr1\nrearrangement\nlength (nt)", "cdr2_rearrangement_length" = "cdr2\nrearrangement\nlength (nt)", "cdr3_rearrangement_length" = "cdr3\nrearrangement\nlength (nt)", "cdr3_length" = "cdr3\nlength (nt)"))

# Create the box plot with facets
ggplot(irae_length_and_germlineness_df, aes(x = severe_irae, y = Value)) +
  geom_boxplot() +
  facet_wrap(~ Category, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Severe irAE", y = "Value")

```

# Histograms instead of boxplots, look at CDR1/2 rearrangement lengths

```{r}

# Create the overlaid histograms with facets and colored by severe_irae
ggplot(irae_length_and_germlineness_df %>% filter(Category %in% c("cdr1_rearrangement_length", "cdr2_rearrangement_length")), aes(x = Value, fill = severe_irae)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(14, 22) +
  scale_fill_manual(values = c("Yes" = "blue", "No" = "red"))

```

# Histograms instead of boxplots, look at CDR3 (rearrangement) lengths

```{r}

# Create the overlaid histograms with facets and colored by severe_irae
ggplot(irae_length_and_germlineness_df %>% filter(Category %in% c("cdr3_rearrangement_length", "cdr3_length")), aes(x = Value, fill = severe_irae)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(0, 80) +
  scale_fill_manual(values = c("Yes" = "blue", "No" = "red"))

```

# Histograms instead of boxplots, look at n1/2 insertions

```{r}

# Create the overlaid histograms with facets and colored by severe_irae
ggplot(irae_length_and_germlineness_df %>% filter(Category %in% c("n1_insertions", "n2_insertions")), aes(x = Value, fill = severe_irae)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(0, 20) +
  scale_fill_manual(values = c("Yes" = "blue", "No" = "red"))

```

# Create df of hydrophobicity scores for amino acids, prepare score calculator

```{r}

# Define the amino acids and their corresponding Eisenberg hydrophobicity values
amino_acids <- c("A", "R", "N", "D", "C", "Q", "E", "G", "H", "I", "L", "K", "M", "F", "P", "S", "T", "W", "Y", "V")
hydrophobicity_values <- c(0.620, -2.530, -0.780, -0.900, 0.290, -0.850, -0.740, 0.480, -0.400, 1.380, 1.060, -1.500, 0.640, 1.190, 0.120, -0.180, -0.050, 0.81, 0.26, 1.08)

# Create a data frame
hydrophobicity_df <- data.frame(Amino_Acid = amino_acids, Hydrophobicity = hydrophobicity_values)

hydrophobicity_df

calculate_hydrophobicity <- function(sequence, hydrophobicity_df) {
  sequence <- toupper(sequence)
  # Calculate the hydrophobicity scores for each amino acid, and sum them
  scores <- sapply(strsplit(sequence, '')[[1]], function(aa) {
    match_aa <- match(aa, hydrophobicity_df$Amino_Acid)
    if (!is.na(match_aa)) {
      return(hydrophobicity_df$Hydrophobicity[match_aa])
    } else {
      # If an amino acid is not found, return NA
      return(NA)
    }
  })
  # If there are any NAs in the scores, return NA; otherwise, return the sum
  if (any(is.na(scores))) {
    return(NA)
  } else {
    return(sum(scores))
  }
}

```

# Investigate hydrophobicity as a function of ICI status

```{r}

ici_hydrophobicity_df <- combined_immunoseq_data %>%
  select(amino_acid, cdr1_amino_acid, cdr2_amino_acid, cdr3_amino_acid, ici_status) %>%
  rowwise() %>%
  mutate(amino_acid_score = calculate_hydrophobicity(amino_acid, hydrophobicity_df)) %>%
  mutate(cdr1_amino_acid_score = calculate_hydrophobicity(cdr1_amino_acid, hydrophobicity_df)) %>%
  mutate(cdr2_amino_acid_score = calculate_hydrophobicity(cdr2_amino_acid, hydrophobicity_df)) %>%
  mutate(cdr3_amino_acid_score = calculate_hydrophobicity(cdr3_amino_acid, hydrophobicity_df)) %>%
  ungroup() %>%
  select(-amino_acid, -cdr1_amino_acid, -cdr2_amino_acid, -cdr3_amino_acid) %>%
  gather(key = "Category", value = "Value", -ici_status)

ici_hydrophobicity_df

```

# Plot hydrophobicity scores as a function of ICI status

```{r}

custom_labeller <- as_labeller(c("amino_acid_score" = "Amino\nacid score", "cdr1_amino_acid_score" = "CDR1 amino\nacid score", "cdr2_amino_acid_score" = "CDR2 amino\nacid score", "cdr3_amino_acid_score" = "CDR3 amino\nacid score"))

# Create the box plot with facets
ggplot(ici_hydrophobicity_df, aes(x = ici_status, y = Value)) +
  geom_boxplot() +
  facet_wrap(~ Category, scales = "free_y", labeller = custom_labeller) +
  labs(x = "ici_status", y = "Value")

```

# Histograms instead of boxplots, look at amino acid and CDR3 AA

```{r}

# Create the overlaid histograms with facets and colored by ici_status
ggplot(ici_hydrophobicity_df %>% filter(Category %in% c("amino_acid_score", "cdr3_amino_acid_score")), aes(x = Value, fill = ici_status)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(-10, 10) +
  scale_fill_manual(values = c("pre" = "blue", "on" = "red"))

```

# Histograms instead of boxplots, look at CDR1/2 AA

```{r}

# Create the overlaid histograms with facets and colored by ici_status
ggplot(ici_hydrophobicity_df %>% filter(Category %in% c("cdr1_amino_acid_score", "cdr2_amino_acid_score")), aes(x = Value, fill = ici_status)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(-6, 6) +
  scale_fill_manual(values = c("pre" = "blue", "on" = "red"))

```

# Investigate hydrophobicity as a function of irAE status

```{r}

irae_hydrophobicity_df <- combined_immunoseq_data %>%
  select(amino_acid, cdr1_amino_acid, cdr2_amino_acid, cdr3_amino_acid, severe_irae) %>%
  rowwise() %>%
  mutate(amino_acid_score = calculate_hydrophobicity(amino_acid, hydrophobicity_df)) %>%
  mutate(cdr1_amino_acid_score = calculate_hydrophobicity(cdr1_amino_acid, hydrophobicity_df)) %>%
  mutate(cdr2_amino_acid_score = calculate_hydrophobicity(cdr2_amino_acid, hydrophobicity_df)) %>%
  mutate(cdr3_amino_acid_score = calculate_hydrophobicity(cdr3_amino_acid, hydrophobicity_df)) %>%
  ungroup() %>%
  select(-amino_acid, -cdr1_amino_acid, -cdr2_amino_acid, -cdr3_amino_acid) %>%
  gather(key = "Category", value = "Value", -severe_irae)

irae_hydrophobicity_df

```

# Plot hydrophobicity scores as a function of irAE status

```{r}

custom_labeller <- as_labeller(c("amino_acid_score" = "Amino\nacid score", "cdr1_amino_acid_score" = "CDR1 amino\nacid score", "cdr2_amino_acid_score" = "CDR2 amino\nacid score", "cdr3_amino_acid_score" = "CDR3 amino\nacid score"))

# Create the box plot with facets
ggplot(irae_hydrophobicity_df, aes(x = severe_irae, y = Value)) +
  geom_boxplot() +
  facet_wrap(~ Category, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Severe irAE", y = "Value")

```

# Histograms instead of boxplots, look at amino acid and CDR3 AA

```{r}

# Create the overlaid histograms with facets and colored by ici_status
ggplot(irae_hydrophobicity_df %>% filter(Category %in% c("amino_acid_score", "cdr3_amino_acid_score")), aes(x = Value, fill = severe_irae)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(-10, 10) +
  scale_fill_manual(values = c("Yes" = "blue", "No" = "red"))

```

# Histograms instead of boxplots, look at CDR1/2 AA

```{r}

# Create the overlaid histograms with facets and colored by ici_status
ggplot(irae_hydrophobicity_df %>% filter(Category %in% c("cdr1_amino_acid_score", "cdr2_amino_acid_score")), aes(x = Value, fill = severe_irae)) +
  geom_density(alpha = 0.5, position = "identity") +
  facet_wrap(~ Category, ncol = 2, scales = "free_y", labeller = custom_labeller) +
  labs(x = "Value", y = "Density") +
  xlim(-6, 6) +
  scale_fill_manual(values = c("Yes" = "blue", "No" = "red"))

```

# Test creating csv for stichr thimble input

```{r}

stitchr_df <- combined_immunoseq_data %>%
  select(v_resolved, d_resolved, j_resolved, cdr3_amino_acid) %>%
  mutate(TCR_name = row_number())

stitchr_df

write.table(stitchr_df, file = "../saved_data/immunoseq_for_stitchr.tsv", sep = "\t", row.names = FALSE)

```