---
title: "Distinct mechanisms of mismatch repair deficiency delineate two modes of response to PD-1 immunotherapy in endometrial carcinoma"
output: pdf_document
---

 - data reproduction of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9905265/
 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

#library(Seurat)
#library(ggplot2)
#library(dplyr)
#library(readr)
#library(readxl)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Their code

```{r}

# All cells
data = readRDS("filtdata-all-coarseT_reann.final.rds") #from GEO
# Full Metadata file:
metadata = read.table("filtdata-all-coarseT_reann.final.metadata.txt",sep="\t",header=TRUE,row.names=1)
data@meta.data=metadata

Idents(data) = "cluster_fine"
options(bitmapType='cairo') 	
png('all-umap.png',width=9,height=6,units='in',res=600,bg='transparent')
DimPlot(data,group.by="cluster_fine",cols=c('#845BA6','#FBC74C','#DFDE4B','#EF4875','#F389A7','#F48268','#8EA33B','#D0A836','#524CA8','#A380BA','#AC1F64','#452B59','#4F7EB7'),raster=F)
dev.off()

# T cells
tdat = subset(data,subset=cluster_fine == "T cells")
Idents(tdat) = "finalIdent"
options(bitmapType='cairo')
png('tcells-umap.png',width=9,height=6,units='in',res=600,bg='transparent')
DimPlot(tdat,group.by="finalIdent",cols = c('#395DA7','#72A5D9','#359E48','#125C2F','#91CD88','#A01F6A','#6B8739','#7F7F7F','#9172B3'),raster=F)
dev.off()

# ============================
# plot Violins with key marker genes
library(Seurat)
data = readRDS("filtdata-all-coarseT_reann.final.rds")
# Full Metadata file:
metadata = read.table("filtdata-all-coarseT_reann.final.metadata.txt",sep="\t",header=TRUE,row.names=1)
data@meta.data=metadata

# All cells
data$cluster_fine=factor(data$cluster_fine,levels=rev(c("T Cells","Naive B Cells","Activated B Cells","Memory B Cells","Plasma Cells","CD56 NK","CD16 NK","CD16 Monocytes","Intermediate Monocytes","CD14 Monocytes","cDC","pDC","Hematopoetic Progenitors")))
Idents(data) = "cluster_fine"
DefaultAssay(data) = "RNA"
mygenes = c("CD3D","CD19","MS4A1","CCR7","IGHD","IGKC","IGHG1","MZB1","GNLY","NCAM1","KLRF1","PRF1","FCGR3A","S100A9","CD14","FCER1A","LILRA4","KIT","CD34")
pdf("all-cells-violin.pdf",height=30,width=35,useDingbats=FALSE)
VlnPlot(data,features=mygenes,pt.size=0,log=TRUE,stack=TRUE)
dev.off()

# T cells
tdat = subset(data,subset=cluster_fine == "T cells")

tdat$finalIdent=factor(tdat$finalIdent,levels=rev(c("CD8 Naive","CD8 Activated","CD8 Exhausted","CD4 Naive","CD4 Activated","Treg","Cycling","GDT","MAIT")))
Idents(tdat)="finalIdent"
DefaultAssay(tdat) = "RNA"
mygenes = c("CD3D","CD8A","CD4","CCR7","TCF7","GZMA","TOX","HAVCR2","CTLA4","MKI67","TRDC","TRAV1-2")
pdf("tCoarse-cells-violin.pdf",height=30,width=35,useDingbats=FALSE)
VlnPlot(tdat,features=mygenes,pt.size=0,log=TRUE,stack=TRUE)
dev.off()

```

# Make sure to bring in healthy donor scRNAseq and scTCRseq data from 102/103 pubs, respectively. 102 = NCBIâ€™s dbGaP at accession number phs002222.v1.p1. 103 = NCBI repository SRA PRJNA717310 (https://www.ncbi.nlm.nih.gov/Traces/study/?query_key=2&WebEnv=MCID_6532e860fb78641bf738aacf&o=acc_s%3Aa&s=SRR14076827,SRR14076831,SRR14076832,SRR14076833,SRR14076834,SRR14076836,SRR14076837,SRR14076838,SRR14076839,SRR14076849,SRR14076850,SRR14076851,SRR14076852,SRR14076853,SRR14076854,SRR14076855,SRR14076867,SRR14076866,SRR14076865,SRR14076864,SRR14076863,SRR14076862,SRR14076860,SRR14076861,SRR14076858)

# Seurat scRNAseq QC: >= 400 counts, >= 150 unique expressed genes, <= 20% mito

# Seurat normalization: SCTransform, regress out counts per cell, features per cell, mito gene expr

# Seurat: itegrate datasets, dimensional reduction & clustering (manual analysis of marker genes and automatic SingleR citation 105), don't ever here use TCR V/J genes for dimensional reduction/clusering (or that will dominate how clustering is done)

# Seurat: take out 14 T cell clusters and re-do dimensional reduction/clustering on them alone. make sure to take out high mito reads clusters

# Goal is to reproduce UMAPs in figure 3: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9905265/figure/F3/