---
title: "scRNAseq reveals distinct T cell populations in irAE of ICI"
output: html_notebook
---

 - data reproduction of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9873824/

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(iCellR)
library(tidyr)

```

# Load samples

```{r}

# List of sample names
sample_names <- c(
  "GSM6668925_P1_T", "GSM6668926_P1_TN"
  #,"GSM6668927_P2_T", "GSM6668928_P2_TN", "GSM6668929_P4_T", "GSM6668930_P4_TN", "GSM6668931_P5_T", "GSM6668932_P5_TN",
  #"GSM6668933_P7_T", "GSM6668934_P7_TN", "GSM6668935_P17_T", "GSM6668936_P17_TN"
  #"GSM6668937_P3_T", "GSM6668938_P3_TN", "GSM6668939_P8_T", "GSM6668940_P8_TN",
  #"GSM6668941_P9_T", "GSM6668942_P9_TN", "GSM6668943_P10_T", "GSM6668944_P10_TN",
  #"GSM6668945_P11_T", "GSM6668946_P11_TN", "GSM6668947_P18_T", "GSM6668948_P18_TN",
  #"GSM6668949_P19_T", "GSM6668950_P19_TN", "GSM6668952_P6_TN", "GSM6668951_P6_T",
  #"GSM6668953_P12_T", "GSM6668954_P12_TN", "GSM6668955_P13_T", "GSM6668956_P13_TN",
  #"GSM6668957_P14_T", "GSM6668958_P14_TN", "GSM6668959_P15_T", "GSM6668960_P15_TN",
  #"GSM6668961_P16_T", "GSM6668962_P16_TN", "GSM6668963_P20_T", "GSM6668964_P20_TN",
  #"GSM6668965_P21_T", "GSM6668966_P21_TN", "GSM6668967_P22_T", "GSM6668968_P22_TN",
  #"GSM6668969_P23_T", "GSM6668970_P23_TN", "GSM6668971_P24_T", "GSM6668972_P24_TN"
)

# Loop through sample names and load 10x data for each sample
my.data <- list()  # Create a list to store the data for each sample

for (sample_name in sample_names) {
  sample_dir <- paste0("./raw_data/", sample_name, "/")
  my.data[[sample_name]] <- load10x(sample_dir)
}


```

# Import metadata, extract pertinent information for sample conditions

```{r}

skip_rows <- 29  # Skip the first 29 rows
num_rows <- 42   # Read 42 rows (30 to 71)

metadata <- read.table("./raw_data/GSE216329_series_matrix.txt", header = TRUE, sep = "\t", skip = skip_rows, nrows = num_rows)

filtered_metadata <- metadata[11:12, ][2:49]

pattern_1 <- "^disease:\\s+"

# Use gsub to replace the values
gsubbed_filtered_metadata <- data.frame(lapply(filtered_metadata, function(col) {
  gsub(pattern_1, "", col)
}))

pattern_2 <- "^condition:\\s+"

reworked_metadata <- data.frame(lapply(gsubbed_filtered_metadata, function(col) {
  gsub(pattern_2, "", col)
}))

reworked_metadata

# Initialize an empty list to store the paired values
paired_values <- list()

# Iterate through the columns
for (col in 1:ncol(reworked_metadata)) {
  # Combine values from row 1 and row 2 with "_"
  paired_values[[col]] <- paste(reworked_metadata[1, col], reworked_metadata[2, col], sep = "_")
}

# Print the list of paired values
print(paired_values)

```

# Aggregate data from all samples

```{r}

sample1 <- my.data$GSM6668925_P1_T
sample2 <- my.data$GSM6668926_P1_TN

# very clunky for now

my.data_aggregate <- data.aggregation(samples = c("sample1","sample2"),
	condition.names = paired_values[0:2])

```

# Convert aggregate data into iCellR object

```{r}

my.obj <- make.obj(my.data_aggregate)

my.obj

```

# Perform QC

```{r}

my.obj <- qc.stats(my.obj)

```

# Visualize QC via violin plots

```{r}

stats.plot(my.obj,
	plot.type = "three.in.one",
	out.name = "UMI-plot",
	interactive = FALSE,
	cell.color = "slategray3", 
	cell.size = 1, 
	cell.transparency = 0.5,
	box.color = "red",
	box.line.col = "green")

```

# Visualize QC via scatter plots

```{r}

stats.plot(my.obj, plot.type = "point.mito.umi", out.name = "mito-umi-plot")
stats.plot(my.obj, plot.type = "point.gene.umi", out.name = "gene-umi-plot")

```

# Filter cells

```{r}

# 33694 cells present before filtering
dim(my.obj@main.data)

my.obj <- cell.filter(my.obj,
	min.mito = 0,
	max.mito = 0.2,
	min.genes = 500,
	max.genes = Inf,
	min.umis = 0,
	max.umis = Inf)

# 11387 cells present after filtering
dim(my.obj@main.data)

```

# Normalize data

```{r}

my.obj <- norm.data(my.obj, 
     norm.method = "ranked.glsf",
     top.rank = 500) # best for scRNA-Seq

```

# Look at gene stats

```{r}

my.obj <- gene.stats(my.obj, which.data = "main.data")

head(my.obj@gene.data[order(my.obj@gene.data$numberOfCells, decreasing = T),])

```

# Make a gene model for clustering

```{r}

# See model plot 
make.gene.model(my.obj, my.out.put = "plot",
	dispersion.limit = 1.5, 
	base.mean.rank = 1500, 
	no.mito.model = T, 
	mark.mito = T, 
	interactive = F,
	out.name = "gene.model")
	
# Write the gene model data into the object

my.obj <- make.gene.model(my.obj, my.out.put = "data",
	dispersion.limit = 1.5, 
	base.mean.rank = 1500, 
	no.mito.model = T, 
	mark.mito = T, 
	interactive = F,
	out.name = "gene.model")

head(my.obj@gene.model)

```

# Run PCA

```{r}

# run PCA in case no batch alignment is necessary
my.obj <- run.pca(my.obj, method = "gene.model", gene.list = my.obj@gene.model,data.type = "main")

opt.pcs.plot(my.obj)

```

# Create UMAP

```{r}

# UMAP
my.obj <- run.umap(my.obj, dims = 1:10)

```