---
title: "scRNAseq reveals distinct T cell populations in irAE of ICI"
output: html_notebook
---

 - data reproduction of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9873824/

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(TAR)
library(Seurat)
library(Matrix)
library(patchwork)
library(ggplot2)

```

# Download data

```{r}

# Specify the URL of the .tar file
url <- "https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE216329&format=file"

# Specify the local file path where you want to save the .tar file
local_file <- "./data/GSE216329.tar"

# Check if the file already exists
if (!file.exists(local_file)) {
  # If the file does not exist, download it
  download.file(url, destfile = local_file, method = "auto")
  cat("Download complete.\n")
} else {
  cat("File already exists. No need to download.\n")
}

```

# Extract contents of the .tar file into the data directory

```{r}

tar_file <- "./data/GSE216329.tar"

# Extract the contents of the .tar file
output_dir <- "./data"

# Use the system command (runs terminal commands from within R) to extract the files, -C specifies directory
system(paste("tar -xf", tar_file, "-C", output_dir))

# manually renamed 2 files (.features. and .enes. to .genes.)

```

# Create a Seurat object for each sample

```{r}

# List of sample names
sample_names <- c(
  "GSM6668925_P1_T", "GSM6668926_P1_TN", "GSM6668927_P2_T", "GSM6668928_P2_TN",
  "GSM6668929_P4_T", "GSM6668930_P4_TN", "GSM6668931_P5_T", "GSM6668932_P5_TN",
  "GSM6668933_P7_T", "GSM6668934_P7_TN", "GSM6668935_P17_T", "GSM6668936_P17_TN",
  "GSM6668937_P3_T", "GSM6668938_P3_TN", "GSM6668939_P8_T", "GSM6668940_P8_TN",
  "GSM6668941_P9_T", "GSM6668942_P9_TN", "GSM6668943_P10_T", "GSM6668944_P10_TN",
  "GSM6668945_P11_T", "GSM6668946_P11_TN", "GSM6668947_P18_T", "GSM6668948_P18_TN",
  "GSM6668949_P19_T", "GSM6668950_P19_TN", "GSM6668952_P6_TN", "GSM6668951_P6_T",
  "GSM6668953_P12_T", "GSM6668954_P12_TN", "GSM6668955_P13_T", "GSM6668956_P13_TN",
  "GSM6668957_P14_T", "GSM6668958_P14_TN", "GSM6668959_P15_T", "GSM6668960_P15_TN",
  "GSM6668961_P16_T", "GSM6668962_P16_TN", "GSM6668963_P20_T", "GSM6668964_P20_TN",
  "GSM6668965_P21_T", "GSM6668966_P21_TN", "GSM6668967_P22_T", "GSM6668968_P22_TN",
  "GSM6668969_P23_T", "GSM6668970_P23_TN", "GSM6668971_P24_T", "GSM6668972_P24_TN"
)

# Directory containing data files
data_dir <- "./data/"

# Create Seurat objects for each sample
seurat_objects <- list()

for (sample_name in sample_names) {
  # Construct file paths
  counts_file <- paste0(data_dir, sample_name, ".matrix.mtx.gz")
  barcodes_file <- paste0(data_dir, sample_name, ".barcodes.tsv.gz")
  genes_file <- paste0(data_dir, sample_name, ".genes.tsv.gz")
  
  # Read data files
  counts <- readMM(counts_file)
  barcodes <- read.table(barcodes_file, stringsAsFactors = FALSE)[, 1]
  features <- read.csv(genes_file, stringsAsFactors = FALSE, sep = "\t", header = FALSE)
  # Remove underscores from feature names and make them unique
  feature_names <- make.unique(gsub("_", "-", features[, 2]))
  rownames(counts) <- feature_names
  colnames(counts) <- barcodes
  
  # Create Seurat object
  seurat <- CreateSeuratObject(counts, project = sample_name)
  
  # Store Seurat object in the list
  seurat_objects[[sample_name]] <- seurat
}


```

# Normalize and identify variable features for each patient separately

```{r}

seurat_objects <- lapply(X = seurat_objects, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

```

# Select features that are repeatedly variable across datasets for integration

```{r}

# some issue with the 27th element, but also trying [-27] doesn't work...

features_1 <- SelectIntegrationFeatures(seurat_objects[1:26])

features_2 <- SelectIntegrationFeatures(seurat_objects[28:length(seurat_objects)])

features <- c(features_1, features_2)

```

# Merge patient sample Seurat objects by finding anchors between patient samples

```{r}

immune.anchors <- FindIntegrationAnchors(object.list = seurat_objects, anchor.features = features)

seurat <- IntegrateData(anchorset = immune.anchors)

```

# Calculate percent mitochondrial, ribosomal, hemoglobin features

```{r}

seurat[["percent.mt"]] <- PercentageFeatureSet(seurat, pattern = "^MT-")
seurat[["percent.rb"]] <- PercentageFeatureSet(seurat, pattern = "^RP[SL]")
seurat[["percent.hb"]] <- PercentageFeatureSet(seurat, pattern = "HB[^(P)]")

```

# Filter out cells based on feature number and percent michondrial features

```{r}

seurat <- subset(seurat, subset = nFeature_RNA > 500 & percent.mt < 20)

```

# Normalize data

```{r}

seurat <- NormalizeData(seurat)

```

# Select features

```{r}

seurat <- FindVariableFeatures(seurat, nfeatures = 2000)

```

# Scale the data

```{r}

seurat <- ScaleData(seurat, vars.to.regress = c("nFeature_RNA", "percent.mt"))

```

# Linear dimensionality reduction using principal component analysis (PCA)

```{r}

seurat <- RunPCA(seurat, features = VariableFeatures(object = seurat), npcs = 50)

ElbowPlot(seurat, ndims = ncol(Embeddings(seurat, "pca")))

```

# See which genes are contributing most to each PC

```{r}

PCHeatmap(seurat, dims = 1:20, cells = 500, balanced = TRUE, ncol = 4)

```
# Visualize PCA

```{r}

DimPlot(seurat, reduction = "pca", raster = FALSE) +
  theme(legend.position = "none")

```

# Non-linear dimension reduction for visualization

```{r}

seurat <- RunUMAP(seurat, dims = 1:20)

```

# Visualize UMAP with patient samples annotated

```{r}

plot1 <- UMAPPlot(seurat, raster = FALSE) +
  theme(legend.position = "none")

plot1

```
# Cluster the cells

```{r}

seurat <- FindNeighbors(seurat, dims = 1:20)

seurat <- FindClusters(seurat, resolution = 1)

```

# Visualize UMAP with cell clusters numbered

```{r}

plot1 <- DimPlot(seurat, reduction = "umap", label = TRUE, raster = FALSE) +
  theme(legend.position = "none")

plot1

```

# Use established markers to determine identities of cell clusters

```{r}

# take markers from https://www.cell.com/cms/10.1016/j.xcrm.2022.100868/attachment/c21c9923-ade8-4a28-8518-aa1f7bc0d1e5/mmc1 table S2

```

# Create KNetL plot

```{r}

# use zoom of 500 (as they used)

```

# Save the resulting Seurat object

```{r}

saveRDS(seurat, file="../data/seurat_obj_all.rds")
#saveRDS(seurat_dorsal, file="../data/seurat_obj_dorsal.rds")

#seurat <- readRDS("DS1/seurat_obj_all.rds")
#seurat_dorsal <- readRDS("DS1/seurat_obj_dorsal.rds")

#save(seurat, seurat_dorsal, file="../data/seurat_objs.rdata")
#load("DS1/seurat_objs.rdata")

```

# Check for batch effects in UMAP, color by variables in metadata (donor, sequencing date... confounders of gene expression, visually check want to see even mixing in each cluster vs. all of one person's/date's samples in one cluster would be bad)