---
title: "P576-1 - ITN ATAC study"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries

```{r}

library(Rsamtools)
library(ggplot2)
library(dplyr)
library(GenomicAlignments)
library(Rsubread)
library(ATACseqQC)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(soGGi)
library(stringr)

```

# Set theme

```{r}

theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill = NA, size = 1),
          axis.text = element_text(colour="black"),
          axis.ticks = element_line(colour="black"),
          axis.text.x = element_text(angle=0),
          legend.title = element_text(size = 10),
          legend.text = element_text(size = 7)))

```

# Excluding failed QC samples, merge bam files by study group. difficult to deal with different peak widths across samples, Erin did consensus peaks (which IS WRONG, this means fixed list of peaks that are all 400 bp wide doesn't take into account longer ones, looks for peak overlaps b/w samples, kind of works but is dangerous because it washes out signal). INSTEAD combine data and instead of consensus peaks MERGE data manually into one big bam file per study group (pile up reads)
# in conda samtools_env environment
# merging (bed/sam tools) goes well COMMAND LINE, but keep code here for completeness

conda init
conda activate samtools_env
samtools merge ./output_file.bam /path/to/input_file_1.bam /path/to/input_file_2.bam etc.

# Re-do peak calling on those merged bam files) to get new peaks (WAY better peaks, more representative of what we're looking for)
# in conda base environment
conda init
macs2 callpeak -t input.bam -g hs -f BAMPE -n name_of_output
# no need for .ending for output name

# THEN merge these (narrow) peak files together into 1 big merged peaks file for visualization on genome browser (will just use narrow peaks/study group for Diffbind so no merging needed for that I don't think)

# merging the narrow peak files across study groups: example code
cat CD4_flaring_peaks.narrowPeak CD4_healthy_peaks.narrowPeak CD4_non-flaring_peaks.narrowPeak > CD4.narrowPeak
sort -k1,1 -k2,2n CD4.narrowPeak > CD4_sorted.bed
bedtools merge -i CD4_sorted.bed > CD4_merged.bed

# Re-run QC script with new input, not expecting to throw anything out (because that'd be throwing out a study group at this point) but I guess it's a good sanity check?